---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

## Mapgen

<!-- badges: start -->
<!-- badges: end -->

Mapgen is a multi-function software that performs the following tasks: 

  1. Enrichment analysis of functional annotations for a trait of interest.
  2. Functionally-informed genetic fine-mapping.
  3. Gene mapping based on fine-mapping result and genomic annotations.

## Installation

You can install the development version of `mapgen` from [GitHub][mapgen-github] with:

```{r install-mapgen-package, eval=FALSE}
install.packages("remotes")
remotes::install_github("xinhe-lab/mapgen")
```

After installing, check that it loads properly:
```{r load-mapgen, eval=FALSE}
library(mapgen)
```

## Tutorials

### 1. [Data preparation][data-preparation-tutorial]
  
  Prepare input data: GWAS summary statistics, LD reference panel, etc.
  
### 2. [Enrichment analysis][enrichment-tutorial]
  
  Assess the enrichment of genetic signals of a trait of interest in functional 
  annotations using `TORUS`.
  
  *Please install [TORUS](https://github.com/xqwen/torus) software package,
if you need to run enrichment analysis. 

### 3. [Fine-mapping][finemapping-tutorial]
  
  Perform Bayesian statistical fine-mapping using `SuSiE` on trait-associated loci, 
  using a informative prior that favors variants located in enriched annotations. 
  
  *Please install [susieR](https://github.com/stephenslab/susieR) package, 
  if you need to run fine-mapping with GWAS summary statistics. 

### 4. [Gene mapping][gene-mapping-tutorial]
  
  Infer causal genes at each locus based on fine-mapping result and genomic 
  annotations, including gene annotations, chromatin loops, etc. 

```{r workflow.overview, include=FALSE, echo=FALSE, fig.cap="", out.width = '80%'}
knitr::include_graphics("man/figures/workflow.overview.png")
```

## Reference

> Alan Selewa\*, Kaixuan Luo\*, Michael Wasney, Linsin Smith, Xiaotong Sun, 
Chenwei Tang, Heather Eckart, Ivan Moskowitz, Anindita Basu, Xin He, 
Sebastian Pott. Single-cell genomics improves the discovery of risk variants 
and genes of atrial fibrillation. _Nat Commun._ 2023 Aug 17;14(1):4999. 
doi: 10.1038/s41467-023-40505-5. PMID: 37591828; PMCID: PMC10435551.



[mapgen-github]: https://github.com/xinhe-lab/mapgen
[tutorials]: https://xinhe-lab.github.io/mapgen/articles/index.html
[data-preparation-tutorial]: https://xinhe-lab.github.io/mapgen/articles/data_preparation_tutorial.html
[enrichment-tutorial]: https://xinhe-lab.github.io/mapgen/articles/enrichment_tutorial.html
[finemapping-tutorial]: https://xinhe-lab.github.io/mapgen/articles/finemapping_tutorial.html
[gene-mapping-tutorial]: https://xinhe-lab.github.io/mapgen/articles/gene_mapping_tutorial.html
[partition-pip-tutorial]: https://xinhe-lab.github.io/mapgen/articles/partition_pip_tutorial.html

