---
title: "Tutorial for data preparation"
author: Kaixuan Luo
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

In this tutorial, we prepare input data for the analysis.

Load the package. 
```{r load-package}
library(mapgen)
```

## Reference genotype panel

If you need to run fine-mapping, you would need to prepare a `bigSNP` object 
for the reference genotype panel, which will be used for computing LD matrices.
You can skip this if you only want to run enrichment analysis or gene mapping 
with precomputed fine-mapping result.

We provided a `bigSNP` object of the reference genotype panel for 
the 1000 Genome (1KG) European population. Otherwise, you could obtain 
your own reference genotypes in PLINK format (bed/bim/fam) 
and use `readBed()` from the `bigsnpr` package to create the `.rds` file.

You need to install the `bigsnpr` package, and use `snp_attach` to load `.rds` file.
```{r load-bigSNP, eval=FALSE, message=FALSE}
library(bigsnpr)
bigSNP <- bigsnpr::snp_attach(rdsfile = 'bigsnpr/EUR_variable_1kg.rds')
```

## Reference LD blocks

We provided within the package the LD blocks from 1000 Genome (1KG) 
European population (in hg19). 

This loads a data frame called `LD_Blocks`. It has four columns: 
chromosome, start and end positions (in hg19), and 
the indices of the LD blocks.
```{r load-LD-blocks}
data('EUR_LD_Blocks', package='mapgen')
head(LD_Blocks, 3)
```

## GWAS summary statistics

The GWAS summary statistics file could be in either tab or comma delimited. 
You could read it using the `vroom` package.

Let's load a small example GWAS dataset (`test.sumstats.txt.gz`) included in the package.
```{r load-gwas-sumstats}
# Loading GWAS summary statistics
library(vroom)
gwas.file <- system.file('extdata', 'test.sumstats.txt.gz', package='mapgen')
sumstats <- vroom::vroom(gwas.file, col_names = TRUE, show_col_types = FALSE)
as.data.frame(sumstats[1:3,] )
```

We need the following columns in the summary statistics:

* sumstats: A data frame of the GWAS summary statistics
* chr: chromosome
* pos: position (*note: the LD reference data included in the package are in hg19!*)  
* beta: effect size (if you have Odds Ratio, you will need to transform it to log(Odds Ratio))  
* se: standard error (SE)  
* a0: reference allele (optional)
* a1: association/effect allele (optional)
* snp: SNP ID (rsID, or chr:position:a1:a2)
* pval: p-value

We run `process_gwas_sumstats()` to load and process the summary statistics.
This cleans GWAS summary statistics and adds information from bigSNP and LD_Blocks (loaded above).
```{r process-gwas-sumstats, eval=FALSE}
gwas.sumstats <- process_gwas_sumstats(sumstats, 
                                       chr='chr', 
                                       pos='position_b37', 
                                       beta='bcac_onco2_beta', 
                                       se='bcac_onco2_se',
                                       a0='a0', 
                                       a1='a1', 
                                       snp='phase3_1kg_id', 
                                       pval='bcac_onco2_P1df_Wald',
                                       LD_Blocks=LD_Blocks,
                                       bigSNP=bigSNP)
```

You can skip bigSNP and/or LD_Blocks if you only need to run enrichment analysis.

Check that the output is cleaned and has appropriate columns:
```{r include=FALSE}
gwas.sumstats <- readRDS(system.file('extdata', 'test.processed.sumstats.rds', package='mapgen'))
```

```{r print-gwas-sumstats}
head(gwas.sumstats, 3)
```

[enrichment-tutorial]: https://xinhe-lab.github.io/mapgen/articles/enrichment_finemapping_tutorial.html
[finemapping-tutorial]: https://xinhe-lab.github.io/mapgen/articles/enrichment_finemapping_tutorial.html
[gene-mapping-tutorial]: https://xinhe-lab.github.io/mapgen/articles/gene_mapping_tutorial.html
