---
title: "Tutorial for data preparation"
author: Kaixuan Luo
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval=TRUE)
```

In this tutorial, we prepare input data for the analysis.

Load the package. 
```{r load-package}
library(mapgen)
```

## Reference panel

Fine-mapping requires linkage-disequilibrium (LD) information. 
We utilize the R package `bigsnpr` to read in PLINK files (bed/bim/fam) into R 
on the fly. If you have PLINK files, you can use `bigsnpr::snp_readBed()` to 
obtain a file with `.rds` extension. 
This RDS file is used for downstream analyses when attached with 
`bigsnpr::snp_attach()`.

_* You can skip this if you only need to run enrichment analysis or 
downstream analysis (e.g. gene mapping) with precomputed fine-mapping result._

We provided a `bigSNP` object of the reference genotype panel from 
the 1000 Genomes (1KG) European population. 

If you are in the He lab at UChicago, you can load the following file from RCC.
Otherwise, you could obtain 
your own reference genotypes in PLINK format (bed/bim/fam) 
and use `readBed()` from the `bigsnpr` package to create the `.rds` file.
```{r load-bigSNP, eval=FALSE, message=FALSE}
library(bigsnpr)
bigSNP <- snp_attach(rdsfile = '/project2/xinhe/1kg/bigsnpr/EUR_variable_1kg.rds')
```

## LD blocks

For the LD blocks, we need a data frame with four columns: 
chromosome, start and end positions, and 
the indices of the LD blocks.

We provided the LD blocks from `LDetect` for the 1000 Genomes (1KG) European population (in hg19).
```{r load-LD-blocks}
LD_blocks <- readRDS(system.file('extdata', 'LD.blocks.EUR.hg19.rds', package='mapgen'))
head(as.data.frame(LD_blocks), 3)
```

_* You can skip this if you only need to run enrichment analysis._

## GWAS summary statistics

Let's load a small example GWAS dataset (`test.sumstats.txt.gz`).
```{r load-gwas-sumstats}
# Read in GWAS summary statistics using the vroom package
library(vroom)
gwas.file <- system.file('extdata', 'test.sumstats.txt.gz', package='mapgen')
sumstats <- vroom(gwas.file, col_names = TRUE, show_col_types = FALSE)
head(as.data.frame(sumstats),3)
```

We need the following columns in the summary statistics:

* sumstats: A data frame of the GWAS summary statistics
* chr: chromosome
* pos: position  
* beta: effect size (if you have Odds Ratio, you will need to transform it to log(Odds Ratio))  
* se: standard error (SE)  
* a0: reference allele
* a1: association/effect allele
* snp: SNP ID (rsID, or chr:position:a1:a2)
* pval: p-value

We run `process_gwas_sumstats()` to load and process the summary statistics.
This cleans GWAS summary statistics, match SNPs with the reference panel 
from the `bigSNP` object, and add locus ID from `LD_Blocks`.
```{r process-gwas-sumstats, eval=FALSE}
gwas.sumstats <- process_gwas_sumstats(sumstats, 
                                       chr='chr', 
                                       pos='position_b37', 
                                       beta='bcac_onco2_beta', 
                                       se='bcac_onco2_se',
                                       a0='a0', 
                                       a1='a1', 
                                       snp='phase3_1kg_id', 
                                       pval='bcac_onco2_P1df_Wald',
                                       LD_Blocks=LD_blocks,
                                       bigSNP=bigSNP)
```

_* You do not need `bigSNP` and `LD_Blocks` if you only need to run enrichment analysis._

Check that the output is cleaned and has appropriate columns:
```{r include=FALSE}
gwas.sumstats <- readRDS(system.file('extdata', 'test.processed.sumstats.rds', package='mapgen'))
```

```{r print-gwas-sumstats}
head(as.data.frame(gwas.sumstats),3)
```

[enrichment-tutorial]: https://xinhe-lab.github.io/mapgen/articles/enrichment_finemapping_tutorial.html
[finemapping-tutorial]: https://xinhe-lab.github.io/mapgen/articles/enrichment_finemapping_tutorial.html
[gene-mapping-tutorial]: https://xinhe-lab.github.io/mapgen/articles/gene_mapping_tutorial.html
