---
title: "Tutorial for enrichment analysis and fine-mapping with functional priors"
author: Kaixuan Luo, Alan Selewa
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
```

Welcome to this tutorial where we perform enrichment analysis and fine-mapping on a small GWAS dataset. 

In this example, we use a genotype reference panel from European population (1KG). 

If you are in the He lab, you can load the `.rds` file from RCC. 
Otherwise, you will need to obtain your own genotypes in PLINK format (bed/bim/fam) 
and use `bigsnpr::readBed()` to create the `.rds` file below.
```{r, message=FALSE}
bigSNP <- bigsnpr::snp_attach(rdsfile = '/project2/xinhe/1kg/bigsnpr/EUR_variable_1kg.rds')
```

Here we show a quick run-through of the enrichment and fine-mapping procedure using a test summary statistics.

Load the package
```{r}
library(mapgen)
```

First we run `run_gwas_cleaner()`, which takes our summary statistics in either tab or comma delimited. 
It also takes a character vector of the 8 column names needed. 
The columns needed have to be given in the following order:  
  
* chromosome (just the number, no "chr", hg19!)  
* position (base pair position, hg19!)  
* beta (if you have Odds Ratio, you will need to transform it to log(Odds Ratio))  
* standard error (SE)  
* reference allele (A,C,T,G)  
* association/effect allele (A,C,T,G)  
* some kind of SNP ID (rsID, or chr:position:a1:a2)  
* p-value  

We also need reference LD blocks. Here we loaded the reference LD blocks from European 1KG. 
```{r}
# load LD blocks
data('Euro_LD_Chunks', package='mapgen')
```

Cleans GWAS summary statistics, adds metadata
```{r}
sumstats_file <- system.file("extdata", "test_sumstats.txt.gz", package="mapgen")
gwas <- run_gwas_cleaner(sumstats_file, 
                         cols.to.keep = c('chr','position_b37','bcac_onco2_beta','bcac_onco2_se','a0','a1','phase3_1kg_id','bcac_onco2_P1df_Wald'),
                         bigSNP = bigSNP,
                         LD_Blocks = LD_Blocks)
```

Check that the output is cleaned and has appropriate columns:
```{r}
gwas[,c('chr','pos','beta','se','snp','pval','zscore')]
```


Next, we must perform enrichment analysis with TORUS, 
which requires annotation files in `.bed` format. 
Use the `prepare_torus_input_files()` function, which takes the previous output, 
and annotation bed files. 
Your bed file should contain only 3 columns (chr, start, end) and 
chromosomes should be just the number (no "chr"). 

*Currently only hg19/b37 coordinates are accepted. You will get wrong results if you use hg38/other.*

```{r}
bed_annotations_dir <- system.file("extdata", "test_bed_dir/", package="mapgen")
annotation_bed_files <- list.files(path = bed_annotations_dir, pattern = "*.bed", full.names = TRUE)
torus.files <- prepare_torus_input_files(gwas, annotation_bed_files)
```

Now that the appropriate files have been generated, lets run TORUS. 

`run_torus()` with `option = "est_prior"` returns a list with: 
enrichment estimates (log odds ratio) and 95% confidence intervals of each annotation, 
and SNP-level priors using the enrichment estimates.
```{r}
torus.result <- run_torus(torus.files$torus_annot_file, 
                          torus.files$torus_zscore_file,
                          option = "est_prior",
                          torus_path = "torus") # you may need to set the installed path to 'torus' executable.
```

TORUS also gives us the uncertainty associated with whether each chunk contains a causal variant or not. 
We run `run_torus()` with `option = "fdr"` to get the probability of each chunk containing a causal variant. 
We can filter our chunks with low probability, thus saving us on computational time!
```{r}
torus.fdr <- run_torus(torus.files$torus_annot_file, 
                       torus.files$torus_zscore_file,
                       option = "fdr",
                       torus_path = "torus")
```

Lets add the TORUS results back onto our cleaned summary statistics. 
`prepare_susie_data()` takes a parameter called `fdr_thresh` which is the FDR associated with each chunk. 
This value is 10% by default, and chunks with FDR < 10% are removed. 
We set it to 1 here just to keep all chunks but you will want to lower this or just use default.

```{r}
susie.df <- prepare_susie_data(sumstats = gwas, torus_pip = torus.result$snp_pip, torus_fdr = torus.fdr, fdr_thresh = 1)
```

We see we have a new column called `torus_pip` 

```{r}
susie.df[,c('chr','pos','beta','se','snp','pval','zscore','torus_pip')]
```

With this data frame, we can perform fine-mapping. We use `RunFinemapping()` to get out results. 
This will be quite slow if chunks contain many SNPs( O(n^2) where n is # of SNPs). 

```{r run-susie}
susie_finemap_L1 <- run_finemapping(sumstats = susie.df, bigSNP = bigSNP)
```

`susie_finemap_L1` is a list of SuSiE results, one for each chunk/LD block. 
Usually we are just interested in the SuSiE PIP, 
which gives the probability of a SNP being causal. 
We can annotate our cleaned summary statistics with this information using `merge_susie_sumstats()`

```{r}
gwas_finemapped <- merge_susie_sumstats(susie_results = susie_finemap_L1, sumstats = susie.df)
```

Lets look at the final cleaned, and finemapped summary statistics. 
We see we have a new column called `susie_pip` which is the probability of being causal. 
Note that we ran SuSiE with L = 1 here, meaning we assumed there is at most 1 causal variant per SNP. 

```{r}
gwas_finemapped[,c('chr','pos','beta','se','snp','pval','zscore','torus_pip','susie_pip')]
```

