---
title: "Tutorial for enrichment analysis and fine-mapping with functional priors"
author: Alan Selewa, Kaixuan Luo
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
```

In this tutorial, we perform enrichment analysis and fine-mapping on a small GWAS dataset.

We use a reference genotype panel from European population (1KG). If you are in the He lab, 
you can load the `EUR_variable_1kg.rds` file from RCC. 

```{r, eval=FALSE, message=FALSE}
bigSNP <- bigsnpr::snp_attach(rdsfile = '/project2/xinhe/1kg/bigsnpr/EUR_variable_1kg.rds')
```

Otherwise, you will need to obtain your own genotypes in PLINK format (bed/bim/fam) 
and use `bigsnpr::readBed()` to create the `.rds` file below.

Load the package
```{r load-package}
library(mapgen)
```

First we run `process_gwas_sumstats()` to load and process the summary statistics.
The summary statistics file could be in either tab or comma delimited.
We need the following columns in the summary statistics:  

* chromosome
* position (base pair position, *we current only support hg19!*)  
* beta (if you have Odds Ratio, you will need to transform it to log(Odds Ratio))  
* standard error (SE)  
* reference allele
* association/effect allele
* SNP ID (rsID, or chr:position:a1:a2)  
* p-value

Load the reference LD blocks from European 1KG. 
```{r load-LD-blocks}
data('Euro_LD_Chunks', package='mapgen')
```

Cleans GWAS summary statistics, adds metadata.
```{r process-gwas}
gwas_file <- system.file("extdata", "test_sumstats.txt.gz", package="mapgen")
gwas <- process_gwas_sumstats(gwas_file, 
                              chr='chr', pos='position_b37', 
                              beta='bcac_onco2_beta', se='bcac_onco2_se',
                              a0='a0', a1='a1', 
                              snp='phase3_1kg_id', pval='bcac_onco2_P1df_Wald',
                              bigSNP=bigSNP,
                              LD_Blocks=LD_Blocks)
```

Check that the output is cleaned and has appropriate columns:
```{r print-gwas}
gwas[,c('chr','pos','beta','se','snp','pval','zscore')]
```


Next, we must perform enrichment analysis with TORUS, 
which requires annotation files in `.bed` format. 
Use the `prepare_torus_input_files()` function, which takes the previous output, 
and annotation bed files. 
Your bed file should contain only 3 columns (chr, start, end) and 
chromosomes should be just the number (no "chr"). 

*Currently only hg19/b37 coordinates are accepted. You will get wrong results if you use hg38/other.*

```{r prepare-torus-files}
bed_annotations_dir <- system.file("extdata", "test_bed_dir/", package="mapgen")
annotation_bed_files <- list.files(path = bed_annotations_dir, pattern = "*.bed", full.names = TRUE)
torus.files <- prepare_torus_input_files(gwas, annotation_bed_files, torus_input_dir = "./torus_input")
```

Now that the appropriate files have been generated, let's run TORUS. 

`run_torus()` with `option = "est-prior"` returns a list with: 
enrichment estimates (log odds ratio) and 95% confidence intervals of each annotation, 
and SNP-level priors using the enrichment estimates.
If you only want to enrichment estimates without computing SNP-level priors, 
you can set `option = "est"`.
```{r run-torus-est-prior}
torus.result <- run_torus(torus.files$torus_annot_file, 
                          torus.files$torus_zscore_file,
                          option = "est-prior",
                          torus_path = "torus") # set the path to 'torus' executable.

torus.enrich <- torus.result$enrich
torus.prior <- torus.result$snp_prior
```

TORUS also gives us the uncertainty of whether each locus contains a causal variant or not. 
We run `run_torus()` with `option = "fdr"` to get the probability of each locus containing a causal variant. 
```{r run-torus-fdr}
torus.fdr <- run_torus(torus.files$torus_annot_file, 
                       torus.files$torus_zscore_file,
                       option = "fdr",
                       torus_path = "torus")$fdr
```

Lets add the TORUS SNP-level priors to our summary statistics. 
```{r add-torus-priors}
sumstats.for.susie <- prepare_susie_data_with_torus_result(sumstats = gwas, torus_prior = torus.prior)
```

We see we have a new column called `torus_prior`.
```{r}
sumstats.for.susie[,c('chr','pos','beta','se','snp','pval','zscore','torus_prior')]
```

You could limit loci with GWAS p-value cutoff (e.g. pval < 5e-8).

If you want to limit loci with FDR cutoff, you can set the parameter `fdr_thresh`, 
which is the FDR associated with each locus obtained from running `run_torus()` with `option = "fdr"`. 
We set it to 1 here just to keep all loci. You could lower this or just use default (FDR < 10%).
```{r add-torus-priors-fdr}
sumstats.for.susie <- prepare_susie_data_with_torus_result(sumstats = gwas, torus_prior = torus.prior, torus_fdr = torus.fdr, fdr_thresh = 1)
```

With this data frame, we can perform fine-mapping. 
We use `run_finemapping()` with SNP-level priors computed from TORUS based on enrichment estimates.
Note that we ran SuSiE with L = 1 here, meaning we assumed there is at most 1 causal variant per SNP. 
This step could take a while if chunks contain many SNPs ($O(n^2)$ where n is # of SNPs). 

```{r run-susie}
susie_finemap_L1 <- run_finemapping(sumstats = sumstats.for.susie, bigSNP = bigSNP, 
                                    priortype = 'torus', L = 1)
```

`susie_finemap_L1` is a list of SuSiE results, one for each chunk/LD block. 
Usually we are just interested in the SuSiE PIP, 
which gives the probability of a SNP being causal. 
We can annotate our cleaned summary statistics with this information using `merge_susie_sumstats()`

```{r}
finemap.sumstats <- merge_susie_sumstats(susie_results = susie_finemap_L1, sumstats = sumstats.for.susie)
```

Lets look at the final cleaned, and finemapped summary statistics. 
We see we have a new column called `susie_pip` which is the probability of being causal. 

```{r}
finemap.sumstats[,c('chr','pos','beta','se','snp','pval','zscore','torus_prior','susie_pip')]
```

