---
title: "Tutorial for fine-mapping with functional priors"
author: Kaixuan Luo, Alan Selewa
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval=TRUE)
```

In this tutorial, we perform fine-mapping (using SuSiE) with 
functional priors computed from the enrichment analysis using `TORUS`.

Load the packages. 
```{r load-package}
library(mapgen)
```

## Input data

Please see the [data preparation tutorial][data-preparation-tutorial] about 
preparing GWAS summary statistics. 

We need a `bigSNP` object for the reference genotype panel, 
which will be used for computing LD matrices. See the 
[data preparation tutorial][data-preparation-tutorial] for details.

## SNP-level priors

Following the [enrichment analysis tutorial][enrichment-tutorial], 
we can obtain SNP-level priors using `run_torus()` with `option = "est-prior"`.
```{r}
gwas.sumstats <- readRDS(system.file('extdata', 'test.processed.sumstats.rds', package='mapgen'))
torus.res <- readRDS(system.file('extdata', 'test.torus.res.rds', package='mapgen'))
```

```{r get-torus-prior}
torus.prior <- torus.res$snp_prior
```

Now, we can add the SNP-level priors to our summary statistics. 
```{r add-torus-priors}
gwas.sumstats <- prepare_susie_data_with_torus_result(sumstats = gwas.sumstats, 
                                                      torus_prior = torus.prior)
```

We see we have a new column called torus_prior.
```{r}
as.data.frame(gwas.sumstats[1:5,])
```

We could limit loci with GWAS p-value cutoff (e.g. 5e-8), 
or if you want to limit loci with FDR cutoff, you can use the 
FDR associated with each locus obtained from running `run_torus()` 
with `option = "fdr"`. 

## Fine-mapping

We can then perform fine-mapping using SuSiE, 
you will need the `susieR` and `bigsnpr` packages for the steps below. 
```{r, message=FALSE}
library(susieR)
library(bigsnpr)
```

To use SNP-level priors computed from `TORUS`, we use `run_finemapping()` 
with `priortype = 'torus'`.
```{r run-susie-torus-prior, eval=FALSE, message=FALSE}
susie.finemap.res <- run_finemapping(sumstats = gwas.sumstats, 
                                     bigSNP = bigSNP, 
                                     priortype = 'torus', 
                                     L = 1)
```

You can also run fine-mapping using uniform prior (without SNP-level priors)
with `priortype = 'uniform'`. (You don't need to run `TORUS` if you run 
fine-mapping using uniform prior.)

Note that we ran SuSiE with L = 1 here, meaning we assumed there is at most 
one causal variant per SNP. This is conservative to avoid potential inflation 
of signals due to mismatch between GWAS and the LD reference.

`susie.finemap.res` is a list of `SuSiE` results, one for each locus. 

We can merge GWAS summary statistics with SuSiE result:
```{r merge-susie-sumstats, eval=FALSE}
finemap.sumstats <- merge_susie_sumstats(susie_results = susie.finemap.res, 
                                         sumstats = gwas.sumstats)
```

We now have a new column called `susie_pip` which is the probability of a SNP 
being causal estimated using `SuSiE`.

```{r, include=FALSE}
finemap.sumstats <- readRDS(system.file('extdata', 'test.finemap.sumstats.rds', package='mapgen'))
```

```{r}
as.data.frame(finemap.sumstats[1:5,])
```

[data-preparation-tutorial]: https://xinhe-lab.github.io/mapgen/articles/data_preparation_tutorial.html
[enrichment-tutorial]: https://xinhe-lab.github.io/mapgen/articles/enrichment_finemapping_tutorial.html
