---
title: "Gene mapping tutorial"
author: Kaixuan Luo, Alan Selewa
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Gene mapping procedure
=================================

Despite our fine-mapping efforts, there remains considerable uncertainty of 
causal variants in most loci. Even if the causal variants are known, 
assigning target genes can be difficult due to complex, 
often long-range, regulatory relationships between enhancers and genes.

We developed a novel gene mapping procedure to prioritize target genes:

1. For every putative causal SNP, we assign a weight to each nearby gene, 
taking into account multiple ways a SNP may affect a gene. 
The weight of a nearby gene can be viewed as the probability 
that this SNP targets that gene. 

2. The PIP of each SNP is then distributed among all potential target genes 
according to the weights of these genes. The “fractional PIP” a gene receives 
from a SNP can be viewed as the support the SNP provides to that gene. 

3. For each gene, we then sum over the fractional PIPs it receives from 
all candidate SNPs in the region. The resulting “gene PIP” approximates the 
probability of a gene being causal. 

4. Similar to variant-level fine-mapping, 
we also define a “credible gene set”, the set of genes that 
capture the causal signal at a locus with high probability (80% by default).

```{r gene-mapping-diagram, echo=FALSE, fig.cap="Schematic of gene-level PIP calculation. s: SNP, g: gene", out.width = '75%'}
knitr::include_graphics("../man/figures/gene.mapping.diagram.png")
```

Gene mapping tutorial
=====================

Our gene mapping procedure takes variant-level PIPs from genetic fine-mapping 
(probability for a SNP being causal), together 
with functional annotation data, and computes gene-level PIPs 
(probability for a gene being causal).

**Required input data**:

* Genetic fine-mapping summary statistics.
* Gene annotations (exons, introns, UTRs, etc.).
* Functional annotation data, including: pcHiC links, ABC scores, 
open chromatin regions (OCRs), enhancer regions (histone ChIP-seq peaks), etc. 

Load R packages
```{r load-packages, message=FALSE, warning=FALSE}
suppressMessages(library(data.table))
suppressMessages(library(tidyverse))
suppressMessages(library(GenomicRanges))
library(ggplot2)
library(mapgen)
```

Load fine-mapping results, filter SNPs with PIP threshold of 1e-5, and save as a GRanges object.
```{r load-finemapping-res}
finemapstats <- readRDS(system.file("extdata", "aFib_Finemapped.tble.rds", package = "mapgen"))
finemapstats.gr <- process_finemapping_sumstats(finemapstats, 
                                                snp = 'snp', chr = 'chr', 
                                                pos = 'pos', pip = 'susie_pip', 
                                                pval = 'pval', zscore = 'zscore', 
                                                cs = 'CS', locus = 'locus',  
                                                pip.thresh = 1e-5)
head(finemapstats.gr)
```

Load genomic annotations (hg19) and gene information.
```{r load-annotations}
genomic.annots <- readRDS(system.file("extdata", "genomic.annots.hg19.gr.rds", package = "mapgen"))
gene.annots <- genomic.annots$genes
```

Load OCRs 
```{r load-OCRs}
genomic.annots$OCRs_hg19 <- readRDS(system.file("extdata", "OCRs.hg19.gr.rds", package = "mapgen"))
```

Define active promoters using OCRs. 
```{r active.promoters}
genomic.annots$active_promoters <- IRanges::subsetByOverlaps(genomic.annots$promoters, genomic.annots$OCRs_hg19, minoverlap = 100)
```

Define enhancer loops: 

Add Promoter-capture HiC (PCHi-C) data from iPSC derived cardiomyocytes (CMs).
```{r load-pcHiC}
genomic.annots$pcHiC <- readRDS(system.file("extdata", "pcHiC.CM.gr.rds", package = "mapgen"))
```

You can use your own pcHiC data as follows, 
or skip this if you do not have relevant PCHi-C data.
```{r load-pcHiC-2, eval=FALSE}
evapcHiC <- data.table::fread(pcHiC.file)
pcHiC.gr <- process_pcHiC(pcHiC)
pcHiC.gr <- pcHiC.gr[pcHiC.gr$gene_name %in% gene.annots$gene_name,] # restrict to protein coding genes
genomic.annots$pcHiC <- pcHiC.gr
```

Add ABC scores from heart ventricle (from Nasser *et al.* *Nature* 2021). 
You may skip this if you do not have relevant ABC scores.
```{r process_ABC}
ABC <- data.table::fread(system.file("extdata", "heart_ventricle-ENCODE_ABC.tsv.gz", package = "mapgen"))
ABC.gr <- process_ABC(ABC, ABC.thresh = 0.015, full.element = TRUE)
ABC.gr <- ABC.gr[ABC.gr$gene_name %in% gene.annots$gene_name, ] # restrict to protein coding genes
genomic.annots$ABC <- ABC.gr
```

Considering the fact that Hi-C and PC-HiC may miss contacts between close regions
due to technical reasons, 
we also consider enhancer regions (the OCRs in this case) within 20 kb of active promoters as “enhancer loops”. 

```{r load-enhancers}
genomic.annots$enhancer_regions <- genomic.annots$OCRs_hg19[genomic.annots$OCRs_hg19$peakType!="Promoter",]
```

```{r enhancer.nearby.promoter.20kb}
enhancer_nearby_promoter_20kb <- nearby_interactions(genomic.annots$enhancer_regions, genomic.annots$active_promoters, dist_limit = 20000)
```

Run gene mapping
===================

Run gene mapping using the following settings: 

* enhancer.loop.method = "ABC.pcHiC.nearby20kb" (as explained above).
* intron.mode = FALSE (do not assign intronic SNPs to the genes containing the introns).
* c.dist = 50000 (distance weight parameter: 50kb).

```{r run-gene-mapping, message=FALSE, warning=FALSE}
gene.mapping.res <- compute_gene_pip(finemapstats.gr, 
                                     genomic.annots,
                                     enhancer.loop.method = "ABC.pcHiC.nearby20kb",
                                     intron.mode = FALSE,
                                     c.dist = 50000)
head(gene.mapping.res)
```

Extract gene-level result table
```{r extract_gene_level_result}
gene.pip.res <- extract_gene_level_result(gene.mapping.res, gene.annots)
head(gene.pip.res)
```

Obtain 80% credible gene sets
```{r gene-cs}
gene.cs.df <- gene_cs(gene.mapping.res, by.locus = TRUE, gene.cs.percent.thresh = 0.8)
head(gene.cs.df)
```

Find the nearest gene for the top SNP in each locus
```{r nearest-genes}
finemapstats.gr <- finemapstats.gr[order(abs(finemapstats.gr$zscore), decreasing = TRUE), ]
topsnps.gr <- finemapstats.gr[!duplicated(finemapstats.gr$locus), ]
  
nearest_genes.df <- find_nearest_genes(topsnps.gr, gene.annots, dist.to = "genebody")
head(nearest_genes.df)
```


Make a Manhattan plot of the gene PIPs, and label genes with gene PIP > 0.8.
```{r gene-manhattan-plot, fig.width=8, fig.height=5, warning=FALSE}
gene_manhattan_plot(gene.pip.res, sig.pip = 0.8, title = "AF gene mapping")
```

