---
title: "Partition fine-mapping PIPs by annotation categories"
author: Kaixuan Luo
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval=TRUE, comment = "#")
```

Partitioning PIPs by cell types and annotation categories
============================================================

Here, we show an example using data from our heart single-cell study.
We assigned the likely cell type(s) through 
which the causal variants act in each locus using fine-mapped SNPs,
and cell-type specific open chromatin regions (OCRs).

**Required input data**:

* Fine-mapping summary statistics.
* A list of GRanges objects of annotation regions. 

Load R packages
```{r load-packages, message=FALSE, warning=FALSE}
library(mapgen)
```

Load fine-mapping summary statistics.
```{r load-finemapping-res}
finemapstats <- readRDS(system.file("extdata", "AF.finemapping.result.rds", package = "mapgen"))
finemapstats <- process_finemapping_sumstats(finemapstats, 
                                             snp = 'snp', chr = 'chr', 
                                             pos = 'pos', pip = 'susie_pip', 
                                             pval = 'pval', zscore = 'zscore', 
                                             cs = 'CS', locus = 'locus',  
                                             pip.thresh = 1e-5)
```

We can partition PIPs into different functional annotation categories.

Load genomic annotations (hg19). 
```{r load-genomic-annotations}
genomic.annots <- readRDS(system.file("extdata", "genomic.annots.hg19.gr.rds", package = "mapgen"))
```

Load all OCRs (hg19).
```{r load-all-OCRs}
genomic.annots$OCRs <- readRDS(system.file("extdata", "OCRs.hg19.gr.rds", package = "mapgen"))
```

Create a list of the annotations, with priority in the order of OCRs, UTRS, Exons, and Introns.
```{r create-annots.list}
annots.list <- list(OCRs = genomic.annots$OCRs,
                    UTRs = genomic.annots$UTRs,
                    Exons = genomic.annots$exons,
                    Introns = genomic.annots$introns)
```

Sum PIPs within annotation categories. 
Unlike partition_pip_regions(), it is OK to have overlapping annotations here. 
If a SNP is in multiple annotation categories, it will be assigned to the first ordered category.
```{r sum-pip-annots}
sum.pip.res <- partition_pip_annots(finemapstats, annots.list)
```

Sum of PIPs in each annotation category:
```{r}
sum.pips <- sum.pip.res$sum.pips
head(sum.pips)
```

Number of SNPs included in each annotation category:
```{r}
n.snps <- sum.pip.res$n.snps
head(n.snps)
```

```{r get-prop-pips-annots}
locus.order <- rownames(sum.pips)[with(sum.pips, order(-OCRs, UTRs, Exons, Introns, others))]
sum.pips <- sum.pips[locus.order,]
prop.pip.mat <- sum.pips/rowSums(sum.pips)
```

We can make a structure plot to show the proportion of PIPs in each annotation category.
```{r structure-plot-annots, fig.width=10, fig.height=2.5}
categories <- c("OCRs", "UTRs", "Exons", "Introns", "others")
colors <- c(OCRs = "#E18727FF", UTRs = "#238b45", Exons =  "#bee6af", Introns = "#B09C85FF", others = "#aaaaaa")
make_structure_plot(prop.pip.mat, categories = categories, colors = colors)
```

We can further partition PIPs into disjoint OCRs for different cell types.

Load cell type OCR peaks. 
This is a list of GRanges objects containing disjoint OCRs for different cell types. 
```{r load-disjoint-OCRs}
disjoint.OCRs <- readRDS(system.file("extdata", "disjoint.OCRs.hg19.grlist.rds", package = "mapgen"))
```

Sum PIPs within cell-type specific OCRs.
```{r sum-pip-disjoint-OCRs}
sum.pip.res <- partition_pip_regions(finemapstats, disjoint.OCRs)
```

Sum of PIPs in each cell type OCR category:
```{r}
sum.pips <- sum.pip.res$sum.pips
head(sum.pips)
```

Filter loci with a cutoff of total PIPs in OCR > 0.25, 
compute the proportion of PIPs partitioned in each cell type category.
```{r filter-locus-proportion-pips-OCRs}
# reorder the loci to match the previous figure
sum.pips <- sum.pips[locus.order, ]
# filter loci with a cutoff of total PIPs in OCR > 0.25
sum.pips.filtered <- sum.pips[rowSums(sum.pips) > 0.25,]
prop.pip.mat <- sum.pips.filtered/rowSums(sum.pips.filtered)
```

We can make a structure plot to show the proportion of PIPs in each cell type category.
```{r structure-plot-OCRs, fig.width=10, fig.height=2.5}
categories <- c("Cardiomyocyte", "Endothelial", "Fibroblast", "Lymphoid", 
                "Myeloid", "Pericyte", "Shared 2-3", "Shared 4+")
colors <- c("#b22222", "#8DD3C7", "#BEBADA", "#FB8072", 
            "#80B1D3", "#B3DE69", "royalblue", "#003C86")
structure_plot(prop.pip.mat, categories = categories, colors = colors)
```
