---
title: "Partition fine-mapping PIPs by annotation categories"
author: Kaixuan Luo
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Partitioning PIPs by cell types and annotation categories
============================================================

Here, we show an example using data from our heart single-cell study.
We assigned the likely cell type(s) through 
which the causal variants act in each locus using fine-mapped SNPs,
and cell-type specific open chromatin regions (OCRs).

**Required input data**:

  * Fine-mapping summary statistics.
  * A list of annotation regions (cell-type specific OCRs in this example). 

Load R packages
```{r load-packages, message=FALSE, warning=FALSE}
suppressMessages(library(tidyverse))
suppressMessages(library(GenomicRanges))
library(Mapgen)
```

Load fine-mapping summary statistics.
```{r load-finemapping-res}
finemap <- readRDS(system.file("extdata", "aFib_Finemapped.tble.rds", package = "Mapgen"))
finemap.gr <- process_finemapping_sumstat(finemap, snp = 'snp', chr = 'chr', 
                                          pos = 'pos', pip = 'susie_pip', 
                                          pval = 'pval', zscore = 'zscore', 
                                          cs = 'CS', locus = 'locus',  
                                          pip.thresh = 1e-5,
                                          cols.to.keep = c('snp','chr','pos', 'pip', 'pval','locus'))
head(finemap.gr)
```

Load genomic annotations (hg19). 
```{r load-annotations}
genomic.annots <- readRDS(system.file("extdata", "genomic.annots.hg19.gr.rds", package = "Mapgen"))
```

Load annotations and cell type OCR peaks (disjoint regions). This is a list of GRanges objects.
```{r load-disjoint-OCRs}
OCRs <- readRDS(system.file("extdata", "disjoint_OCRs_hg19.grlist.rds", package = "Mapgen"))
summary(OCRs)
```

Sum of PIPs within cell-type specific OCRs.
```{r sum-pip-disjoint-OCRs}
sum_pip_res <- partition_pip_regions(finemap.gr, OCRs)
sum.pips <- sum_pip_res$sum.pips
n.snps <- sum_pip_res$n.snps

cat('Sum of PIPs in cell-type specific OCRs\n')
head(sum.pips, 3)

cat('Number of SNPs included:\n')
head(n.snps, 3)

```

Filter loci with a cutoff of total PIPs in OCR > 0.25, 
compute the proportion of PIPs partitioned in each cell type category.
```{r filter-locus-proportion-pips}
sum.pips.filtered <- sum.pips[rowSums(sum.pips) > 0.25,]

prop.pip.mat <- sum.pips.filtered/rowSums(sum.pips.filtered)

prop.pip.mat <- prop.pip.mat[with(prop.pip.mat, order(-Cardiomyocyte)), ]

```

We can make a structure plot to show the proportion of PIPs in each cell type category.
```{r structure-plot, fig.width=10, fig.height=2.5}
categories <- c("Cardiomyocyte", "Endothelial", "Fibroblast", "Lymphoid", 
                "Myeloid", "Pericyte", "Shared 2-3", "Shared 4+")
colors <- c("Cardiomyocyte" = "#b22222", "Endothelial" = "#8DD3C7",
            "Fibroblast" = "#BEBADA", "Lymphoid" = "#FB8072", 
            "Myeloid" = "#80B1D3", "Pericyte" = "#B3DE69", 
            "Shared 2-3" = "royalblue", "Shared 4+" = "#003C86")
dat <- compile_structure_plot_data(prop.pip.mat, categories = rev(categories))

structure_plot(dat, colors = rev(colors))
```
