<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Tutorial for data preparation • mapgen</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Tutorial for data preparation">
<meta property="og:description" content="mapgen">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mapgen</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.5.1.9001</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/data_preparation_tutorial.html">Data preparation</a>
    </li>
    <li>
      <a href="../articles/enrichment_tutorial.html">Enrichment analysis</a>
    </li>
    <li>
      <a href="../articles/finemapping_tutorial.html">Functional fine-mapping</a>
    </li>
    <li>
      <a href="../articles/partition_pip_tutorial.html">Partition fine-mapping PIPs by annotation categories</a>
    </li>
    <li>
      <a href="../articles/gene_mapping_tutorial.html">Gene mapping</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/xinhe-lab/mapgen/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Tutorial for data preparation</h1>
                        <h4 data-toc-skip class="author">Kaixuan
Luo</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/mapgen/blob/HEAD/vignettes/data_preparation_tutorial.Rmd" class="external-link"><code>vignettes/data_preparation_tutorial.Rmd</code></a></small>
      <div class="hidden name"><code>data_preparation_tutorial.Rmd</code></div>

    </div>

    
    
<p>In this tutorial, we prepare input data for the analysis.</p>
<p>Load the package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/xinhe-lab/mapgen" class="external-link">mapgen</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="reference-genotype-panel">Reference genotype panel<a class="anchor" aria-label="anchor" href="#reference-genotype-panel"></a>
</h2>
<p>If you need to run fine-mapping, you would need to prepare a
<code>bigSNP</code> object for the reference genotype panel, which will
be used for computing LD matrices. You can skip this if you only want to
run enrichment analysis or gene mapping with precomputed fine-mapping
result.</p>
<p>We provided a <code>bigSNP</code> object of the reference genotype
panel for the 1000 Genome (1KG) European population. Otherwise, you
could obtain your own reference genotypes in PLINK format (bed/bim/fam)
and use <code>readBed()</code> from the <code>bigsnpr</code> package to
create the <code>.rds</code> file.</p>
<p>You need to install the <code>bigsnpr</code> package, and use
<code>snp_attach</code> to load <code>.rds</code> file.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://privefl.github.io/bigsnpr/" class="external-link">bigsnpr</a></span><span class="op">)</span></span>
<span><span class="va">bigSNP</span> <span class="op">&lt;-</span> <span class="fu">bigsnpr</span><span class="fu">::</span><span class="fu"><a href="https://privefl.github.io/bigsnpr/reference/snp_attach.html" class="external-link">snp_attach</a></span><span class="op">(</span>rdsfile <span class="op">=</span> <span class="st">'bigsnpr/EUR_variable_1kg.rds'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="reference-ld-blocks">Reference LD blocks<a class="anchor" aria-label="anchor" href="#reference-ld-blocks"></a>
</h2>
<p>We provided within the package the LD blocks from 1000 Genome (1KG)
European population (in hg19).</p>
<p>This loads a data frame called <code>LD_Blocks</code>. It has four
columns: chromosome, start and end positions (in hg19), and the indices
of the LD blocks.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">'EUR_LD_Blocks'</span>, package<span class="op">=</span><span class="st">'mapgen'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">LD_Blocks</span>, <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 3 × 4</span></span></span>
<span><span class="co">##      X1      X2      X3    X4</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span>     1   <span style="text-decoration: underline;">10</span>583 1<span style="text-decoration: underline;">892</span>607     1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span>     1 1<span style="text-decoration: underline;">892</span>607 3<span style="text-decoration: underline;">582</span>736     2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span>     1 3<span style="text-decoration: underline;">582</span>736 4<span style="text-decoration: underline;">380</span>811     3</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="gwas-summary-statistics">GWAS summary statistics<a class="anchor" aria-label="anchor" href="#gwas-summary-statistics"></a>
</h2>
<p>The GWAS summary statistics file could be in either tab or comma
delimited. You could read it using the <code>vroom</code> package.</p>
<p>Let’s load a small example GWAS dataset
(<code>test.sumstats.txt.gz</code>) included in the package.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Loading GWAS summary statistics</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://vroom.r-lib.org" class="external-link">vroom</a></span><span class="op">)</span></span>
<span><span class="va">gwas.file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'test.sumstats.txt.gz'</span>, package<span class="op">=</span><span class="st">'mapgen'</span><span class="op">)</span></span>
<span><span class="va">sumstats</span> <span class="op">&lt;-</span> <span class="fu">vroom</span><span class="fu">::</span><span class="fu"><a href="https://vroom.r-lib.org/reference/vroom.html" class="external-link">vroom</a></span><span class="op">(</span><span class="va">gwas.file</span>, col_names <span class="op">=</span> <span class="cn">TRUE</span>, show_col_types <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">sumstats</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="op">]</span> <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                           var_name                              phase3_1kg_id</span></span>
<span><span class="co">## 1 1_10616_CCGCCGTTGCAAAGGCGCGCCG_C rs376342519:10616:CCGCCGTTGCAAAGGCGCGCCG:C</span></span>
<span><span class="co">## 2                      1_11008_C_G                                1:11008:C:G</span></span>
<span><span class="co">## 3                      1_11012_C_G                                1:11012:C:G</span></span>
<span><span class="co">##   chr position_b37                     a0 a1 bcac_onco2_r2</span></span>
<span><span class="co">## 1   1        10616 CCGCCGTTGCAAAGGCGCGCCG  C      0.837659</span></span>
<span><span class="co">## 2   1        11008                      C  G      0.580657</span></span>
<span><span class="co">## 3   1        11012                      C  G      0.580657</span></span>
<span><span class="co">##   bcac_onco2_eaf_controls bcac_onco2_beta bcac_onco2_se bcac_onco2_P1df_Wald</span></span>
<span><span class="co">## 1                  0.9925         0.00009       0.05784             0.998700</span></span>
<span><span class="co">## 2                  0.0979        -0.04365       0.02002             0.029280</span></span>
<span><span class="co">## 3                  0.0979        -0.04365       0.02002             0.029279</span></span>
<span><span class="co">##   bcac_onco2_P1df_LRT bcac_icogs2_r2       rsIDs</span></span>
<span><span class="co">## 1            0.998700       0.262460 rs376342519</span></span>
<span><span class="co">## 2            0.029368       0.335425           1</span></span>
<span><span class="co">## 3            0.029367       0.335425           1</span></span></code></pre>
<p>We need the following columns in the summary statistics:</p>
<ul>
<li>sumstats: A data frame of the GWAS summary statistics</li>
<li>chr: chromosome</li>
<li>pos: position (<em>note: the LD reference data included in the
package are in hg19!</em>)<br>
</li>
<li>beta: effect size (if you have Odds Ratio, you will need to
transform it to log(Odds Ratio))<br>
</li>
<li>se: standard error (SE)<br>
</li>
<li>a0: reference allele (optional)</li>
<li>a1: association/effect allele (optional)</li>
<li>snp: SNP ID (rsID, or chr:position:a1:a2)</li>
<li>pval: p-value</li>
</ul>
<p>We run <code><a href="../reference/process_gwas_sumstats.html">process_gwas_sumstats()</a></code> to load and process the
summary statistics. This cleans GWAS summary statistics and adds
information from bigSNP and LD_Blocks (loaded above).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gwas.sumstats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_gwas_sumstats.html">process_gwas_sumstats</a></span><span class="op">(</span><span class="va">sumstats</span>, </span>
<span>                                       chr<span class="op">=</span><span class="st">'chr'</span>, </span>
<span>                                       pos<span class="op">=</span><span class="st">'position_b37'</span>, </span>
<span>                                       beta<span class="op">=</span><span class="st">'bcac_onco2_beta'</span>, </span>
<span>                                       se<span class="op">=</span><span class="st">'bcac_onco2_se'</span>,</span>
<span>                                       a0<span class="op">=</span><span class="st">'a0'</span>, </span>
<span>                                       a1<span class="op">=</span><span class="st">'a1'</span>, </span>
<span>                                       snp<span class="op">=</span><span class="st">'phase3_1kg_id'</span>, </span>
<span>                                       pval<span class="op">=</span><span class="st">'bcac_onco2_P1df_Wald'</span>,</span>
<span>                                       LD_Blocks<span class="op">=</span><span class="va">LD_Blocks</span>,</span>
<span>                                       bigSNP<span class="op">=</span><span class="va">bigSNP</span><span class="op">)</span></span></code></pre></div>
<p>You can skip bigSNP and/or LD_Blocks if you only need to run
enrichment analysis.</p>
<p>Check that the output is cleaned and has appropriate columns:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gwas.sumstats</span>, <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 3 × 12</span></span></span>
<span><span class="co">##     chr   pos a0    a1       beta     se snp           pval zscore locus og_in…¹</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span>     1 <span style="text-decoration: underline;">13</span>110 A     G     -<span style="color: #BB0000;">0.020</span><span style="color: #BB0000; text-decoration: underline;">7</span> 0.032<span style="text-decoration: underline;">3</span> 1:13110:G:A  0.522 -<span style="color: #BB0000;">0.639</span>     1       3</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span>     1 <span style="text-decoration: underline;">13</span>116 G     T     -<span style="color: #BB0000;">0.012</span><span style="color: #BB0000; text-decoration: underline;">6</span> 0.019<span style="text-decoration: underline;">0</span> rs201725126… 0.508 -<span style="color: #BB0000;">0.662</span>     1       4</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span>     1 <span style="text-decoration: underline;">13</span>118 G     A     -<span style="color: #BB0000;">0.012</span><span style="color: #BB0000; text-decoration: underline;">6</span> 0.019<span style="text-decoration: underline;">0</span> rs200579949… 0.508 -<span style="color: #BB0000;">0.662</span>     1       5</span></span>
<span><span class="co">## <span style="color: #949494;"># … with 1 more variable: bigSNP_index &lt;int&gt;, and abbreviated variable name</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   ¹​og_index</span></span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kaixuan Luo, Alan Selewa, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
