<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Functional enrichment and fine mapping with TORUS and SuSiE • Mapgen</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Functional enrichment and fine mapping with TORUS and SuSiE">
<meta property="og:description" content="Mapgen">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Mapgen</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/enrichment_finemapping_tutorial.html">Functional enrichment and fine mapping tutorial</a>
    </li>
    <li>
      <a href="../articles/gene_mapping_tutorial.html">Gene mapping tutorial</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Functional enrichment and fine mapping with TORUS and SuSiE</h1>
                        <h4 data-toc-skip class="author">Alan Selewa, Kaixuan Luo</h4>
            
      
      
      <div class="hidden name"><code>enrichment_finemapping_procedure.Rmd</code></div>

    </div>

    
    
<div class="section level1">
<h1 id="enrichment-analysis-using-torus">Enrichment analysis using TORUS<a class="anchor" aria-label="anchor" href="#enrichment-analysis-using-torus"></a>
</h1>
<p>We perform enrichment analysis with TORUS, which requires annotation files in <code>.bed</code> format.</p>
<p>To generate input files for TORUS, use the <code><a href="../reference/PrepareTorusFiles.html">PrepareTorusFiles()</a></code> function, which takes the GWAS summary statistics, and a directory with bed files. Your bed file should contain only 3 columns (chr, start, end) and chromosomes should be just the number (no “chr”).</p>
<p><em>Currently only hg19/b37 coordinates are accepted. You will get wrong results if you use hg38/other.</em></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">torus.files</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PrepareTorusFiles.html">PrepareTorusFiles</a></span><span class="op">(</span><span class="va">gwas</span>, bed_annotations <span class="op">=</span> <span class="va">bed_annotations</span><span class="op">)</span></code></pre></div>
<p>Now that the appropriate files have been generated (they are in .temp/), let’s run TORUS.</p>
<p><code><a href="../reference/RunTorus.html">RunTorus()</a></code> returns: enrichment level of each annotation in log2 units, and SNP level priors, which will be used as priors for finemapping.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">torus.result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RunTorus.html">RunTorus</a></span><span class="op">(</span>torus_annot_file <span class="op">=</span> <span class="va">torus.files</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, torus_zscore_file <span class="op">=</span> <span class="va">torus.files</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<p>TORUS also gives us the uncertainty associated with whether each chunk contains a causal variant or not.</p>
<p>We run <code><a href="../reference/RunTorusFDR.html">RunTorusFDR()</a></code> to get the probability of each chunk containing a causal variant. We can filter out chunks with low probability, thus saving us on computational time!</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">torus.fdr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RunTorusFDR.html">RunTorusFDR</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level1">
<h1 id="run-susie-with-functional-priors-from-torus">Run SuSiE with functional priors from TORUS<a class="anchor" aria-label="anchor" href="#run-susie-with-functional-priors-from-torus"></a>
</h1>
<p>Prepare a data frame with both GWAS summary statistics and SNP-level priors from TORUS.</p>
<p><code><a href="../reference/PrepareSusieData.html">PrepareSusieData()</a></code> takes a parameter called <code>fdr_thresh</code> which is the FDR associated with each chunk. This value is 10% by default, and chunks with FDR &lt; 10% are removed. (We set it to 1 here just to keep all chunks but you will want to lower this or just use default.)</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">susie.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PrepareSusieData.html">PrepareSusieData</a></span><span class="op">(</span>sumstats <span class="op">=</span> <span class="va">gwas</span>, torus_pip <span class="op">=</span> <span class="va">torus.result</span><span class="op">$</span><span class="va">snp_pip</span>, torus_fdr <span class="op">=</span> <span class="va">torus.fdr</span>, fdr_thresh <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p>With this data frame, we can perform fine mapping using SuSiE.</p>
<p>We use <code><a href="../reference/RunFinemapping.html">RunFinemapping()</a></code> to get out results. This will be quite slow if chunks contain many SNPs( O(n^2) where n is # of SNPs). Note that we ran SuSiE with L = 1 here, meaning we assumed there is at most 1 causal variant per SNP.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">susie_finemap_L1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RunFinemapping.html">RunFinemapping</a></span><span class="op">(</span>sumstats <span class="op">=</span> <span class="va">susie.df</span>, bigSNP <span class="op">=</span> <span class="va">bigSNP</span><span class="op">)</span></code></pre></div>
<p><code>susie_finemap_L1</code> is a list of SuSiE results, one for each chunk/LD block.</p>
<p>Usually we are just interested in the SuSiE PIP, which gives the probability of a SNP being causal. We can annotate our cleaned summary statistics with this information using <code><a href="../reference/merge_susie_sumstats.html">merge_susie_sumstats()</a></code></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gwas_finemapped</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/merge_susie_sumstats.html">merge_susie_sumstats</a></span><span class="op">(</span>susie_results <span class="op">=</span> <span class="va">susie_finemap_L1</span>, sumstats <span class="op">=</span> <span class="va">susie.df</span><span class="op">)</span></code></pre></div>
<p><code>gwas_finemapped</code> gives us fine mapping summary statistics. We will have a column called <code>susie_pip</code> which is the probability for a SNP being causal.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kaixuan Luo, Alan Selewa, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
