<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Tutorial for enrichment analysis and fine-mapping with functional priors • mapgen</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Tutorial for enrichment analysis and fine-mapping with functional priors">
<meta property="og:description" content="mapgen">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mapgen</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.12</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/enrichment_finemapping_tutorial.html">Functional enrichment and fine-mapping</a>
    </li>
    <li>
      <a href="../articles/partition_pip_tutorial.html">Partition fine-mapping PIPs by annotation categories</a>
    </li>
    <li>
      <a href="../articles/gene_mapping_tutorial.html">Gene mapping</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Tutorial for enrichment analysis and
fine-mapping with functional priors</h1>
                        <h4 data-toc-skip class="author">Kaixuan Luo,
Alan Selewa</h4>
            
      
      
      <div class="hidden name"><code>enrichment_finemapping_tutorial.Rmd</code></div>

    </div>

    
    
<p>In this tutorial, we perform enrichment analysis and fine-mapping
(using TORUS + SuSiE) on a small GWAS dataset.</p>
<p>We use a reference genotype panel from European population (1KG).</p>
<p>If you are in the He lab, you can load the
<code>EUR_variable_1kg.rds</code> file from RCC.</p>
<p>We need the <code>bigsnpr</code> package for this tutorial.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://privefl.github.io/bigsnpr/" class="external-link">bigsnpr</a></span><span class="op">)</span></span>
<span><span class="va">bigSNP</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://privefl.github.io/bigsnpr/reference/snp_attach.html" class="external-link">snp_attach</a></span><span class="op">(</span>rdsfile <span class="op">=</span> <span class="st">'/project2/xinhe/1kg/bigsnpr/EUR_variable_1kg.rds'</span><span class="op">)</span></span></code></pre></div>
<p>Otherwise, you will need to obtain your own genotypes in PLINK format
(bed/bim/fam) and use <code>readBed()</code> from the
<code>bigsnpr</code> package to create the <code>.rds</code> file
below.</p>
<p>Load the packages. We need the <code>susieR</code> package to run
finemapping.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">mapgen</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stephenslab/susieR" class="external-link">susieR</a></span><span class="op">)</span></span></code></pre></div>
<p>First we run <code><a href="../reference/process_gwas_sumstats.html">process_gwas_sumstats()</a></code> to load and process
the summary statistics. The summary statistics file could be in either
tab or comma delimited. We need the following columns in the summary
statistics:</p>
<ul>
<li>chromosome</li>
<li>position (base pair position, <em>we current only support
hg19!</em>)<br>
</li>
<li>beta (if you have Odds Ratio, you will need to transform it to
log(Odds Ratio))<br>
</li>
<li>standard error (SE)<br>
</li>
<li>reference allele</li>
<li>association/effect allele</li>
<li>SNP ID (rsID, or chr:position:a1:a2)<br>
</li>
<li>p-value</li>
</ul>
<p>Load the reference LD blocks from European 1KG.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">'Euro_LD_Chunks'</span>, package<span class="op">=</span><span class="st">'mapgen'</span><span class="op">)</span></span></code></pre></div>
<p>Cleans GWAS summary statistics, adds metadata.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gwas_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"test_sumstats.txt.gz"</span>, package<span class="op">=</span><span class="st">"mapgen"</span><span class="op">)</span></span>
<span><span class="va">gwas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_gwas_sumstats.html">process_gwas_sumstats</a></span><span class="op">(</span><span class="va">gwas_file</span>, </span>
<span>                              chr<span class="op">=</span><span class="st">'chr'</span>, pos<span class="op">=</span><span class="st">'position_b37'</span>, </span>
<span>                              beta<span class="op">=</span><span class="st">'bcac_onco2_beta'</span>, se<span class="op">=</span><span class="st">'bcac_onco2_se'</span>,</span>
<span>                              a0<span class="op">=</span><span class="st">'a0'</span>, a1<span class="op">=</span><span class="st">'a1'</span>, </span>
<span>                              snp<span class="op">=</span><span class="st">'phase3_1kg_id'</span>, pval<span class="op">=</span><span class="st">'bcac_onco2_P1df_Wald'</span>,</span>
<span>                              bigSNP<span class="op">=</span><span class="va">bigSNP</span>,</span>
<span>                              LD_Blocks<span class="op">=</span><span class="va">LD_Blocks</span><span class="op">)</span></span></code></pre></div>
<p>Check that the output is cleaned and has appropriate columns:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gwas</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'chr'</span>,<span class="st">'pos'</span>,<span class="st">'beta'</span>,<span class="st">'se'</span>,<span class="st">'snp'</span>,<span class="st">'pval'</span>,<span class="st">'zscore'</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<p>Next, we must perform enrichment analysis with TORUS, which requires
annotation files in <code>.bed</code> format. Use the
<code><a href="../reference/prepare_torus_input_files.html">prepare_torus_input_files()</a></code> function, which takes the
previous output, and annotation bed files. Your bed file should contain
only 3 columns (chr, start, end) and chromosomes should be just the
number (no “chr”).</p>
<p><em>Currently only hg19/b37 coordinates are accepted. You will get
wrong results if you use hg38/other.</em></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bed_annotations_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"test_bed_dir/"</span>, package<span class="op">=</span><span class="st">"mapgen"</span><span class="op">)</span></span>
<span><span class="va">annotation_bed_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">bed_annotations_dir</span>, pattern <span class="op">=</span> <span class="st">"*.bed"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">torus.files</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepare_torus_input_files.html">prepare_torus_input_files</a></span><span class="op">(</span><span class="va">gwas</span>, <span class="va">annotation_bed_files</span>, torus_input_dir <span class="op">=</span> <span class="st">"./torus_input"</span><span class="op">)</span></span></code></pre></div>
<p>Now that the appropriate files have been generated, let’s run
TORUS.</p>
<p><code><a href="../reference/run_torus.html">run_torus()</a></code> with <code>option = "est-prior"</code>
returns a list with: enrichment estimates (log odds ratio) and 95%
confidence intervals of each annotation, and SNP-level priors using the
enrichment estimates. If you only want to enrichment estimates without
computing SNP-level priors, you can set <code>option = "est"</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">torus.result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_torus.html">run_torus</a></span><span class="op">(</span><span class="va">torus.files</span><span class="op">$</span><span class="va">torus_annot_file</span>, </span>
<span>                          <span class="va">torus.files</span><span class="op">$</span><span class="va">torus_zscore_file</span>,</span>
<span>                          option <span class="op">=</span> <span class="st">"est-prior"</span>,</span>
<span>                          torus_path <span class="op">=</span> <span class="st">"torus"</span><span class="op">)</span> <span class="co"># set the path to 'torus' executable.</span></span>
<span></span>
<span><span class="va">torus.enrich</span> <span class="op">&lt;-</span> <span class="va">torus.result</span><span class="op">$</span><span class="va">enrich</span></span>
<span><span class="va">torus.prior</span> <span class="op">&lt;-</span> <span class="va">torus.result</span><span class="op">$</span><span class="va">snp_prior</span></span></code></pre></div>
<p>TORUS also gives us the uncertainty of whether each locus contains a
causal variant or not. We run <code><a href="../reference/run_torus.html">run_torus()</a></code> with
<code>option = "fdr"</code> to get the probability of each locus
containing a causal variant.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">torus.fdr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_torus.html">run_torus</a></span><span class="op">(</span><span class="va">torus.files</span><span class="op">$</span><span class="va">torus_annot_file</span>, </span>
<span>                       <span class="va">torus.files</span><span class="op">$</span><span class="va">torus_zscore_file</span>,</span>
<span>                       option <span class="op">=</span> <span class="st">"fdr"</span>,</span>
<span>                       torus_path <span class="op">=</span> <span class="st">"torus"</span><span class="op">)</span><span class="op">$</span><span class="va">fdr</span></span></code></pre></div>
<p>Lets add the TORUS SNP-level priors to our summary statistics.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sumstats.for.susie</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepare_susie_data_with_torus_result.html">prepare_susie_data_with_torus_result</a></span><span class="op">(</span>sumstats <span class="op">=</span> <span class="va">gwas</span>, </span>
<span>                                                           torus_prior <span class="op">=</span> <span class="va">torus.prior</span><span class="op">)</span></span></code></pre></div>
<p>We see we have a new column called <code>torus_prior</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sumstats.for.susie</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'chr'</span>,<span class="st">'pos'</span>,<span class="st">'beta'</span>,<span class="st">'se'</span>,<span class="st">'snp'</span>,<span class="st">'pval'</span>,<span class="st">'zscore'</span>,<span class="st">'torus_prior'</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<p>You could limit loci with GWAS p-value cutoff (e.g. pval &lt;
5e-8).</p>
<p>If you want to limit loci with FDR cutoff, you can set the parameter
<code>fdr_thresh</code>, which is the FDR associated with each locus
obtained from running <code><a href="../reference/run_torus.html">run_torus()</a></code> with
<code>option = "fdr"</code>. We set it to 1 here just to keep all loci.
You could lower this or just use default (FDR &lt; 10%).</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sumstats.for.susie</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepare_susie_data_with_torus_result.html">prepare_susie_data_with_torus_result</a></span><span class="op">(</span>sumstats <span class="op">=</span> <span class="va">gwas</span>, </span>
<span>                                                           torus_prior <span class="op">=</span> <span class="va">torus.prior</span>, </span>
<span>                                                           torus_fdr <span class="op">=</span> <span class="va">torus.fdr</span>, </span>
<span>                                                           fdr_thresh <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>With this data frame, we can perform fine-mapping using SuSiE, you
will need the <code>susieR</code> package for the steps below.</p>
<p>We use <code><a href="../reference/run_finemapping.html">run_finemapping()</a></code> with SNP-level priors computed
from TORUS based on enrichment estimates. Note that we ran SuSiE with L
= 1 here, meaning we assumed there is at most 1 causal variant per SNP.
This step could take a while if chunks contain many SNPs (<span class="math inline">\(O(n^2)\)</span> where n is # of SNPs).</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">susie_finemap_L1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_finemapping.html">run_finemapping</a></span><span class="op">(</span>sumstats <span class="op">=</span> <span class="va">sumstats.for.susie</span>, </span>
<span>                                    bigSNP <span class="op">=</span> <span class="va">bigSNP</span>, </span>
<span>                                    priortype <span class="op">=</span> <span class="st">'torus'</span>, </span>
<span>                                    L <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><code>susie_finemap_L1</code> is a list of SuSiE results, one for
each chunk/LD block. Usually we are just interested in the SuSiE PIP,
which gives the probability of a SNP being causal. We can annotate our
cleaned summary statistics with this information using
<code><a href="../reference/merge_susie_sumstats.html">merge_susie_sumstats()</a></code></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">finemap.sumstats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/merge_susie_sumstats.html">merge_susie_sumstats</a></span><span class="op">(</span>susie_results <span class="op">=</span> <span class="va">susie_finemap_L1</span>, </span>
<span>                                         sumstats <span class="op">=</span> <span class="va">sumstats.for.susie</span><span class="op">)</span></span></code></pre></div>
<p>Lets look at the final cleaned, and finemapped summary statistics. We
see we have a new column called <code>susie_pip</code> which is the
probability of being causal.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">finemap.sumstats</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'chr'</span>,<span class="st">'pos'</span>,<span class="st">'beta'</span>,<span class="st">'se'</span>,<span class="st">'snp'</span>,<span class="st">'pval'</span>,<span class="st">'zscore'</span>,<span class="st">'torus_prior'</span>,<span class="st">'susie_pip'</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kaixuan Luo, Alan Selewa, Sebastian Pott, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
