<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Gene mapping tutorial • Mapgen</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Gene mapping tutorial">
<meta property="og:description" content="Mapgen">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Mapgen</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/enrichment_finemapping_tutorial.html">Functional enrichment and fine mapping tutorial</a>
    </li>
    <li>
      <a href="../articles/gene_mapping_tutorial.html">Gene mapping tutorial</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Gene mapping tutorial</h1>
                        <h4 data-toc-skip class="author">Kaixuan Luo</h4>
            
      
      
      <div class="hidden name"><code>gene_mapping_tutorial.Rmd</code></div>

    </div>

    
    
<div class="section level1">
<h1 id="about-our-gene-mapping-procedure">About our gene mapping procedure<a class="anchor" aria-label="anchor" href="#about-our-gene-mapping-procedure"></a>
</h1>
<p>Despite our fine mapping efforts, there remains considerable uncertainty of causal variants in most loci. Even if the causal variants are known, assigning target genes can be difficult due to complex, often long-range, regulatory relationships between enhancers and genes.</p>
<p>We developed a novel gene mapping procedure to prioritize target genes:</p>
<ol style="list-style-type: decimal">
<li><p>For every putative causal SNP, we assign a weight to each nearby gene, taking into account multiple ways a SNP may affect a gene. The weight of a nearby gene can be viewed as the probability that this SNP targets that gene.</p></li>
<li><p>The PIP of each SNP is then distributed among all potential target genes according to the weights of these genes. The “fractional PIP” a gene receives from a SNP can be viewed as the support the SNP provides to that gene.</p></li>
<li><p>For each gene, we then sum over the fractional PIPs it receives from all candidate SNPs in the region. The resulting “gene PIP” approximates the probability of a gene being causal (Methods).</p></li>
<li><p>Similar to variant-level fine mapping, we also define a “gene credible set”, the set of genes that capture the causal signal at a locus with high probability (80% in our study)</p></li>
</ol>
<div class="figure">
<img src="gene.mapping.diagram.png" alt="Schematic of gene-level PIP calculation. s: SNP, g: gene" width="75%"><p class="caption">
Schematic of gene-level PIP calculation. s: SNP, g: gene
</p>
</div>
</div>
<div class="section level1">
<h1 id="gene-mapping-tutorial">Gene mapping tutorial<a class="anchor" aria-label="anchor" href="#gene-mapping-tutorial"></a>
</h1>
<p>Our gene mapping procedure takes variant-level PIPs from genetic fine mapping (probability for a SNP being causal), together with functional annotation data, and computes gene-level PIPs (probability for a gene being causal).</p>
<p><strong>Required input data</strong>:</p>
<ul>
<li>Genetic fine mapping result</li>
<li>Gene annotations (exons, introns, UTRs, etc.)</li>
<li>Functional annotation data, including: pcHiC links, ABC scores, open chromatin regions (OCRs), enhancer regions (histone ChIP-seq peaks), etc.</li>
</ul>
<p>Load R packages</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">Mapgen</span><span class="op">)</span></code></pre></div>
<p>Load fine-mapping results:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">finemap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"../inst/extdata/aFib_Finemapped.tble.rds"</span><span class="op">)</span>
<span class="va">finemap.gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_finemapping_sumstat.html">process_finemapping_sumstat</a></span><span class="op">(</span><span class="va">finemap</span>, snp <span class="op">=</span> <span class="st">'snp'</span>, chr <span class="op">=</span> <span class="st">'chr'</span>, 
                                          pos <span class="op">=</span> <span class="st">'pos'</span>, pip <span class="op">=</span> <span class="st">'susie_pip'</span>, 
                                          pval <span class="op">=</span> <span class="st">'pval'</span>, zscore <span class="op">=</span> <span class="st">'zscore'</span>, 
                                          cs <span class="op">=</span> <span class="st">'CS'</span>, locus <span class="op">=</span> <span class="st">'locus'</span>,  
                                          cols.to.keep <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'snp'</span>,<span class="st">'chr'</span>,<span class="st">'pos'</span>, <span class="st">'a0'</span>, <span class="st">'a1'</span>, <span class="st">'pip'</span>, <span class="st">'pval'</span>, <span class="st">'zscore'</span>, <span class="st">'cs'</span>, <span class="st">'locus'</span><span class="op">)</span>,
                                          pip.thresh <span class="op">=</span> <span class="fl">1e-5</span>, filterCS <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Process fine mapping summary statistics ...</span>
<span class="co">## Filter SNPs with PIP threshold of 1e-05</span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="va">finemap.gr</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## GRanges object with 6 ranges and 10 metadata columns:</span>
<span class="co">##       seqnames    ranges strand |         snp       chr       pos          a0</span>
<span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt;</span>
<span class="co">##   [1]     chr9 139082621      * |   rs4842129         9 139082621           A</span>
<span class="co">##   [2]     chr9 139085865      * |  rs11103373         9 139085865           C</span>
<span class="co">##   [3]     chr9 139089679      * |   rs7860634         9 139089679           G</span>
<span class="co">##   [4]     chr9 139089717      * |   rs7850466         9 139089717           T</span>
<span class="co">##   [5]     chr9 139089718      * |   rs7847312         9 139089718           A</span>
<span class="co">##   [6]     chr9 139091460      * |   rs2274114         9 139091460           A</span>
<span class="co">##                a1         pip      pval    zscore        cs     locus</span>
<span class="co">##       &lt;character&gt;   &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span>
<span class="co">##   [1]           G 3.00183e-05   5.14929   4.46914         0      1009</span>
<span class="co">##   [2]           T 1.59627e-05   4.78489   4.32857         0      1009</span>
<span class="co">##   [3]           A 1.94283e-03   6.99568   5.33333         0      1009</span>
<span class="co">##   [4]           G 1.42424e-03   6.94808   5.27143         0      1009</span>
<span class="co">##   [5]           C 1.32523e-03   6.91045   5.25714         0      1009</span>
<span class="co">##   [6]           G 6.54232e-05   5.46750  -4.65909         0      1009</span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 20 sequences from an unspecified genome; no seqlengths</span></code></pre>
<p>Load genomic annotations (prepared earlier for hg19).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">genomic.annots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">'../inst/extdata/genomic.annots.hg19.gr.rds'</span><span class="op">)</span></code></pre></div>
<p>Gene information:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gene.annots</span> <span class="op">&lt;-</span> <span class="va">genomic.annots</span><span class="op">$</span><span class="va">genes</span></code></pre></div>
<p>We already included Promoter-capture HiC (PCHi-C) data from iPSC derived cardiomyocytes (CMs) in the ‘genomic.annots.hg19.gr.rds’ file.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="va">genomic.annots</span><span class="op">$</span><span class="va">pcHiC</span>, <span class="fl">3</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## GRanges object with 3 ranges and 5 metadata columns:</span>
<span class="co">##       seqnames            ranges strand | promoter_chr promoter_start</span>
<span class="co">##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;character&gt;      &lt;numeric&gt;</span>
<span class="co">##   [1]     chr1 10056320-10057012      * |         chr1       10002213</span>
<span class="co">##   [2]     chr1 10062800-10063261      * |         chr1       10002213</span>
<span class="co">##   [3]     chr1 10085205-10085681      * |         chr1       10002213</span>
<span class="co">##       promoter_end   gene_name     score</span>
<span class="co">##          &lt;numeric&gt; &lt;character&gt; &lt;numeric&gt;</span>
<span class="co">##   [1]     10004089      NMNAT1      6.33</span>
<span class="co">##   [2]     10004089      NMNAT1      7.37</span>
<span class="co">##   [3]     10004089      NMNAT1      7.28</span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 24 sequences from an unspecified genome; no seqlengths</span></code></pre>
<p>You can use your own pcHiC data, or skip this if you do not have relevant PCHi-C data.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># pcHiC.gr &lt;- process_pcHiC(pcHiC)</span>
<span class="co"># pcHiC.gr &lt;- pcHiC.gr[pcHiC.gr$gene_name %in% gene.annots$gene_name,] # restrict to protein coding genes</span>
<span class="co"># genomic.annots$pcHiC &lt;- pcHiC.gr</span></code></pre></div>
<p>Add ABC scores from heart ventricle (downloaded from Nasser <em>et al.</em> <em>Nature</em> 2021 paper). You may skip this if you do not have relevant ABC scores.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ABC</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">'../inst/extdata/heart_ventricle-ENCODE_ABC.tsv.gz'</span><span class="op">)</span>
<span class="va">ABC.gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_ABC.html">process_ABC</a></span><span class="op">(</span><span class="va">ABC</span>, ABC.thresh <span class="op">=</span> <span class="fl">0.015</span>, full.element <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">ABC.gr</span> <span class="op">&lt;-</span> <span class="va">ABC.gr</span><span class="op">[</span><span class="va">ABC.gr</span><span class="op">$</span><span class="va">gene_name</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">gene.annots</span><span class="op">$</span><span class="va">gene_name</span>, <span class="op">]</span> <span class="co"># restrict to protein coding genes</span>
<span class="va">genomic.annots</span><span class="op">$</span><span class="va">ABC</span> <span class="op">&lt;-</span> <span class="va">ABC.gr</span>
<span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ABC.gr</span>, <span class="fl">3</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## GRanges object with 3 ranges and 23 metadata columns:</span>
<span class="co">##       seqnames        ranges strand |                   name element_start</span>
<span class="co">##          &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; |            &lt;character&gt;   &lt;character&gt;</span>
<span class="co">##   [1]     chr1 888243-888743      * | genic|chr1:888243-88..        888243</span>
<span class="co">##   [2]     chr1 908361-908861      * | genic|chr1:908361-90..        908361</span>
<span class="co">##   [3]     chr1 908361-908861      * | genic|chr1:908361-90..        908361</span>
<span class="co">##       element_end       class activity_base   gene_name TargetGeneTSS</span>
<span class="co">##       &lt;character&gt; &lt;character&gt;     &lt;numeric&gt; &lt;character&gt;     &lt;integer&gt;</span>
<span class="co">##   [1]      888743       genic      0.643903       NOC2L        894679</span>
<span class="co">##   [2]      908861       genic      2.420474      KLHL17        895966</span>
<span class="co">##   [3]      908861       genic      2.420474     PLEKHN1        901876</span>
<span class="co">##       TargetGeneExpression TargetGenePromoterActivityQuantile</span>
<span class="co">##                  &lt;logical&gt;                          &lt;numeric&gt;</span>
<span class="co">##   [1]                 &lt;NA&gt;                           0.506151</span>
<span class="co">##   [2]                 &lt;NA&gt;                           0.431726</span>
<span class="co">##   [3]                 &lt;NA&gt;                           0.462930</span>
<span class="co">##       TargetGeneIsExpressed  distance isSelfPromoter hic_contact</span>
<span class="co">##                   &lt;logical&gt; &lt;numeric&gt;      &lt;logical&gt;   &lt;numeric&gt;</span>
<span class="co">##   [1]                  TRUE      6186          FALSE    0.071404</span>
<span class="co">##   [2]                  TRUE     12645          FALSE    0.021792</span>
<span class="co">##   [3]                  TRUE      6735          FALSE    0.057115</span>
<span class="co">##       powerlaw_contact powerlaw_contact_reference hic_contact_pl_scaled</span>
<span class="co">##              &lt;numeric&gt;                  &lt;numeric&gt;             &lt;numeric&gt;</span>
<span class="co">##   [1]         0.104071                   0.102573              0.070376</span>
<span class="co">##   [2]         0.055676                   0.055071              0.021556</span>
<span class="co">##   [3]         0.096610                   0.095260              0.056317</span>
<span class="co">##       hic_pseudocount hic_contact_pl_scaled_adj ABC.Score.Numerator ABC.Score</span>
<span class="co">##             &lt;numeric&gt;                 &lt;numeric&gt;           &lt;numeric&gt; &lt;numeric&gt;</span>
<span class="co">##   [1]        0.001216                  0.071592            0.046098  0.015224</span>
<span class="co">##   [2]        0.001216                  0.022771            0.055118  0.017673</span>
<span class="co">##   [3]        0.001216                  0.057533            0.139257  0.041100</span>
<span class="co">##       powerlaw.Score.Numerator powerlaw.Score               CellType</span>
<span class="co">##                      &lt;numeric&gt;      &lt;numeric&gt;            &lt;character&gt;</span>
<span class="co">##   [1]                 0.066047       0.012388 heart_ventricle-ENCODE</span>
<span class="co">##   [2]                 0.133298       0.024471 heart_ventricle-ENCODE</span>
<span class="co">##   [3]                 0.230575       0.038288 heart_ventricle-ENCODE</span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 23 sequences from an unspecified genome; no seqlengths</span></code></pre>
<p>Define active promoters.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">genomic.annots</span><span class="op">$</span><span class="va">active.promoters</span> <span class="op">&lt;-</span> <span class="fu">IRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html" class="external-link">subsetByOverlaps</a></span><span class="op">(</span><span class="va">genomic.annots</span><span class="op">$</span><span class="va">promoters</span>, <span class="va">genomic.annots</span><span class="op">$</span><span class="va">OCRs_hg19</span>, minoverlap <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></code></pre></div>
<p>Add enhancer data (you may skip this if you do not have relevant enhancer data)</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">genomic.annots</span><span class="op">$</span><span class="va">enhancer_regions</span> <span class="op">&lt;-</span> <span class="va">genomic.annots</span><span class="op">$</span><span class="va">OCRs_hg19</span></code></pre></div>
</div>
<div class="section level1">
<h1 id="run-gene-mapping">Run gene mapping<a class="anchor" aria-label="anchor" href="#run-gene-mapping"></a>
</h1>
<p>Definition of enhancer loops: “enhancer loops” are defined from Activity-By-Contact (ABC) scores and promoter-capture HiC data. Considering the fact that enhancers close to promoters may actually be missed by PCHi-C and similar assays, because of the difficulty of calling such interactions due to high background level, we also include OCRs that are within 20 kb of active promoter.</p>
<p>Run gene mapping using the following settings:</p>
<ul>
<li>enhancer_loop_method = “ABC.pcHiC.nearby20kb” (as explained above)</li>
<li>intron.mode = FALSE (do not assign intronic SNPs to the genes containing the introns)</li>
<li>distance weight = 50000 (50kb)</li>
</ul>
<p>SNP view table</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">snp.gene.pip.mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_gene_pip.html">compute_gene_pip</a></span><span class="op">(</span><span class="va">finemap.gr</span>, 
                                    <span class="va">genomic.annots</span>,
                                    enhancer_loop_method <span class="op">=</span> <span class="st">"ABC.pcHiC.nearby20kb"</span>,
                                    intron.mode <span class="op">=</span> <span class="cn">FALSE</span>,
                                    c.dist <span class="op">=</span> <span class="fl">50000</span>, 
                                    dist.to <span class="op">=</span> <span class="st">"tss"</span>,
                                    cols.to.keep <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"snp"</span>,<span class="st">"chr"</span>,<span class="st">"pos"</span>,<span class="st">'a0'</span>, <span class="st">'a1'</span>, <span class="st">"locus"</span>, <span class="st">"cs"</span>, <span class="st">"pip"</span>, <span class="st">"gene_name"</span>, <span class="st">"category"</span>, <span class="st">"weight"</span>, <span class="st">"frac_pip"</span>, <span class="st">"gene_pip"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Map SNPs to genes and assign weights ...</span>
<span class="co">## Include ABC scores in enhancer loops ... </span>
<span class="co">## Include pcHiC in enhancer loops ... </span>
<span class="co">## Include enhancers with nearby promoters (20kb) in enhancer loops ... </span>
<span class="co">## Assigning SNPs to genes...</span>
<span class="co">## Assign SNPs in exons and active promoters ...</span>
<span class="co">## Assign SNPs in enhancers to linked genes through enhancer loops ...</span>
<span class="co">## Assign SNPs in enhancer regions to genes by distance weighting ...</span>
<span class="co">## Assign SNPs in introns/UTRs... </span>
<span class="co">## Assign intergenic SNPs to genes by distance weighting ...</span>
<span class="co">## Assign weights to SNP-gene pairs ... </span>
<span class="co">## Compute gene PIP ...</span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="va">snp.gene.pip.mat</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 6 × 13</span></span>
<span class="co">## <span style="color: #949494;"># Groups:   snp [5]</span></span>
<span class="co">##   snp      chr     pos a0    a1    locus    cs     pip gene_name category weight</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> rs665…     1  2.23<span style="color: #949494;">e</span>7 C     T        15     0 8.75<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span> HSPG2     ABC,nea…      1</span>
<span class="co">## <span style="color: #BCBCBC;">2</span> rs665…     1  2.23<span style="color: #949494;">e</span>7 C     T        15     0 8.75<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span> USP48     ABC           1</span>
<span class="co">## <span style="color: #BCBCBC;">3</span> rs213…     1  4.16<span style="color: #949494;">e</span>7 C     T        25     0 5.17<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span> SCMH1     ABC           1</span>
<span class="co">## <span style="color: #BCBCBC;">4</span> rs484…     1  1.54<span style="color: #949494;">e</span>8 G     T        77     0 1.15<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span> IL6R      ABC           1</span>
<span class="co">## <span style="color: #BCBCBC;">5</span> rs668…     1  1.54<span style="color: #949494;">e</span>8 A     G        77     0 1.39<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span> IL6R      ABC           1</span>
<span class="co">## <span style="color: #BCBCBC;">6</span> rs279…     1  2.06<span style="color: #949494;">e</span>8 T     C       106     0 1.95<span style="color: #949494;">e</span><span style="color: #BB0000;">-4</span> SLC45A3   ABC           1</span>
<span class="co">## <span style="color: #949494;"># … with 2 more variables: frac_pip &lt;dbl&gt;, gene_pip &lt;dbl&gt;</span></span></code></pre>
<p>Gene view table</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gene.pip.res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extract_gene_level_result.html">extract_gene_level_result</a></span><span class="op">(</span><span class="va">snp.gene.pip.mat</span>, <span class="va">gene.annots</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Extract gene level result ...</span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gene.pip.res</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 6 × 4</span></span>
<span class="co">##   chr         pos gene_name  gene_pip</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> chr1   22<span style="text-decoration: underline;">263</span>790 HSPG2     0.470    </span>
<span class="co">## <span style="color: #BCBCBC;">2</span> chr1   22<span style="text-decoration: underline;">110</span>099 USP48     0.021<span style="text-decoration: underline;">3</span>   </span>
<span class="co">## <span style="color: #BCBCBC;">3</span> chr1   41<span style="text-decoration: underline;">707</span>826 SCMH1     0.076<span style="text-decoration: underline;">1</span>   </span>
<span class="co">## <span style="color: #BCBCBC;">4</span> chr1  154<span style="text-decoration: underline;">377</span>669 IL6R      0.000<span style="text-decoration: underline;">033</span>0</span>
<span class="co">## <span style="color: #BCBCBC;">5</span> chr1  205<span style="text-decoration: underline;">649</span>587 SLC45A3   0.047<span style="text-decoration: underline;">2</span>   </span>
<span class="co">## <span style="color: #BCBCBC;">6</span> chr1  205<span style="text-decoration: underline;">719</span>404 NUCKS1    0.678</span></code></pre>
<p>80% Gene credible set</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gene.cs.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gene_cs.html">gene_cs</a></span><span class="op">(</span><span class="va">snp.gene.pip.mat</span>, by.locus <span class="op">=</span> <span class="cn">TRUE</span>, gene.cs.percent.thresh <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gene.cs.df</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 6 × 6</span></span>
<span class="co">##   locus gene_cs    gene_cs_locus_pip      top_gene top_locus_gene_… top_gene_pip</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span>     7 CASZ1      CASZ1(0.886)           CASZ1               0.886        0.886</span>
<span class="co">## <span style="color: #BCBCBC;">2</span>    15 HSPG2,CEL… HSPG2(0.47),CELA3B(0.… HSPG2               0.470        0.470</span>
<span class="co">## <span style="color: #BCBCBC;">3</span>    25 SLFNL1,CT… SLFNL1(0.624),CTPS1(0… SLFNL1              0.624        0.624</span>
<span class="co">## <span style="color: #BCBCBC;">4</span>    31 BEND5      BEND5(0.997)           BEND5               0.997        0.997</span>
<span class="co">## <span style="color: #BCBCBC;">5</span>    32 FAF1,CDKN… FAF1(0.382),CDKN2C(0.… FAF1                0.382        0.403</span>
<span class="co">## <span style="color: #BCBCBC;">6</span>    33 TTC39A,EP… TTC39A(0.409),EPS15(0… TTC39A              0.409        0.411</span></code></pre>
<p>Find the nearest gene for the top SNP in each locus</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">nearest_genes.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/find_nearest_genes.html">find_nearest_genes</a></span><span class="op">(</span><span class="va">finemap.gr</span>, <span class="va">gene.annots</span>, dist.to <span class="op">=</span> <span class="st">"genebody"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="va">nearest_genes.df</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##          snp chr       pos nearest_gene</span>
<span class="co">## 1  rs2200732   4 111710042        PITX2</span>
<span class="co">## 2  rs2106261  16  73051620        ZFHX3</span>
<span class="co">## 3 rs11264280   1 154862952        KCNN3</span>
<span class="co">## 4 rs11598047  10 105342672        NEURL</span>
<span class="co">## 5 rs11773845   7 116191301         CAV1</span>
<span class="co">## 6   rs577676   1 170587340        PRRX1</span></code></pre>
<p>Make a Manhattan plot of the gene PIP, and label genes with gene PIP &gt; 0.8.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/gene_manhattan_plot.html">gene_manhattan_plot</a></span><span class="op">(</span><span class="va">gene.pip.res</span>, sig.pip <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span></code></pre></div>
<p><img src="gene_mapping_tutorial_files/figure-html/gene-manhattan-plot-1.png" width="960"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kaixuan Luo, Alan Selewa, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
