<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Gene mapping tutorial • mapgen</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Gene mapping tutorial">
<meta property="og:description" content="mapgen">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mapgen</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/enrichment_finemapping_tutorial.html">Functional enrichment and fine-mapping</a>
    </li>
    <li>
      <a href="../articles/partition_pip_tutorial.html">Partition fine-mapping PIPs by annotation categories</a>
    </li>
    <li>
      <a href="../articles/gene_mapping_tutorial.html">Gene mapping</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Gene mapping tutorial</h1>
                        <h4 data-toc-skip class="author">Kaixuan Luo, Alan Selewa</h4>
            
      
      
      <div class="hidden name"><code>gene_mapping_tutorial.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="gene-mapping-procedure">Gene mapping procedure<a class="anchor" aria-label="anchor" href="#gene-mapping-procedure"></a>
</h2>
<p>Despite our fine-mapping efforts, there remains considerable uncertainty of causal variants in most loci. Even if the causal variants are known, assigning target genes can be difficult due to complex, often long-range, regulatory relationships between enhancers and genes.</p>
<p>We developed a novel gene mapping procedure to prioritize target genes:</p>
<ol style="list-style-type: decimal">
<li><p>For every putative causal SNP, we assign a weight to each nearby gene, taking into account multiple ways a SNP may affect a gene. The weight of a nearby gene can be viewed as the probability that this SNP targets that gene.</p></li>
<li><p>The PIP of each SNP is then distributed among all potential target genes according to the weights of these genes. The “fractional PIP” a gene receives from a SNP can be viewed as the support the SNP provides to that gene.</p></li>
<li><p>For each gene, we then sum over the fractional PIPs it receives from all candidate SNPs in the region. The resulting “gene PIP” approximates the probability of a gene being causal.</p></li>
<li><p>Similar to variant-level fine-mapping, we also define a “credible gene set”, the set of genes that capture the causal signal at a locus with high probability (80% by default).</p></li>
</ol>
<div class="figure">
<img src="../reference/figures/gene.mapping.diagram.png" alt="Schematic of gene-level PIP calculation. s: SNP, g: gene" width="75%"><p class="caption">
Schematic of gene-level PIP calculation. s: SNP, g: gene
</p>
</div>
</div>
<div class="section level2">
<h2 id="gene-mapping-tutorial">Gene mapping tutorial<a class="anchor" aria-label="anchor" href="#gene-mapping-tutorial"></a>
</h2>
<p>Our gene mapping procedure takes variant-level PIPs from genetic fine-mapping (probability for a SNP being causal), together with functional annotation data, and computes gene-level PIPs (probability for a gene being causal).</p>
<p><strong>Required input data</strong>:</p>
<ul>
<li>Genetic fine-mapping summary statistics.</li>
<li>Gene annotations (exons, introns, UTRs, etc.).</li>
<li>Functional annotation data, including: pcHiC links, ABC scores, open chromatin regions (OCRs), enhancer regions (histone ChIP-seq peaks), etc.</li>
</ul>
<p>Load R packages</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">mapgen</span><span class="op">)</span></code></pre></div>
<p>Load fine-mapping results, filter SNPs with PIP threshold of 1e-5, and save as a GRanges object.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">finemapstats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"aFib_Finemapped.tble.rds"</span>, package <span class="op">=</span> <span class="st">"mapgen"</span><span class="op">)</span><span class="op">)</span>
<span class="va">finemapstats.gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_finemapping_sumstats.html">process_finemapping_sumstats</a></span><span class="op">(</span><span class="va">finemapstats</span>, 
                                                snp <span class="op">=</span> <span class="st">'snp'</span>, chr <span class="op">=</span> <span class="st">'chr'</span>, 
                                                pos <span class="op">=</span> <span class="st">'pos'</span>, pip <span class="op">=</span> <span class="st">'susie_pip'</span>, 
                                                pval <span class="op">=</span> <span class="st">'pval'</span>, zscore <span class="op">=</span> <span class="st">'zscore'</span>, 
                                                cs <span class="op">=</span> <span class="st">'CS'</span>, locus <span class="op">=</span> <span class="st">'locus'</span>,  
                                                pip.thresh <span class="op">=</span> <span class="fl">1e-5</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Processing fine-mapping summary statistics ...</span>
<span class="co">## Filter SNPs with PIP threshold of 1e-05</span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="va">finemapstats.gr</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## GRanges object with 6 ranges and 8 metadata columns:</span>
<span class="co">##       seqnames    ranges strand |         snp       chr       pos         pip</span>
<span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;integer&gt; &lt;integer&gt;   &lt;numeric&gt;</span>
<span class="co">##   [1]     chr9 139082621      * |   rs4842129         9 139082621 3.00183e-05</span>
<span class="co">##   [2]     chr9 139085865      * |  rs11103373         9 139085865 1.59627e-05</span>
<span class="co">##   [3]     chr9 139089679      * |   rs7860634         9 139089679 1.94283e-03</span>
<span class="co">##   [4]     chr9 139089717      * |   rs7850466         9 139089717 1.42424e-03</span>
<span class="co">##   [5]     chr9 139089718      * |   rs7847312         9 139089718 1.32523e-03</span>
<span class="co">##   [6]     chr9 139091460      * |   rs2274114         9 139091460 6.54232e-05</span>
<span class="co">##            pval    zscore        cs     locus</span>
<span class="co">##       &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span>
<span class="co">##   [1]   5.14929   4.46914         0      1009</span>
<span class="co">##   [2]   4.78489   4.32857         0      1009</span>
<span class="co">##   [3]   6.99568   5.33333         0      1009</span>
<span class="co">##   [4]   6.94808   5.27143         0      1009</span>
<span class="co">##   [5]   6.91045   5.25714         0      1009</span>
<span class="co">##   [6]   5.46750  -4.65909         0      1009</span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 20 sequences from an unspecified genome; no seqlengths</span></code></pre>
<p>Load genomic annotations (hg19) and gene information.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">genomic.annots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"genomic.annots.hg19.gr.rds"</span>, package <span class="op">=</span> <span class="st">"mapgen"</span><span class="op">)</span><span class="op">)</span>
<span class="va">gene.annots</span> <span class="op">&lt;-</span> <span class="va">genomic.annots</span><span class="op">$</span><span class="va">genes</span></code></pre></div>
<p>Load OCRs</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">genomic.annots</span><span class="op">$</span><span class="va">OCRs_hg19</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"OCRs.hg19.gr.rds"</span>, package <span class="op">=</span> <span class="st">"mapgen"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Define active promoters using OCRs.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">genomic.annots</span><span class="op">$</span><span class="va">active_promoters</span> <span class="op">&lt;-</span> <span class="fu">IRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html" class="external-link">subsetByOverlaps</a></span><span class="op">(</span><span class="va">genomic.annots</span><span class="op">$</span><span class="va">promoters</span>, <span class="va">genomic.annots</span><span class="op">$</span><span class="va">OCRs_hg19</span>, minoverlap <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></code></pre></div>
<p>Define enhancer loops:</p>
<p>Add Promoter-capture HiC (PCHi-C) data from iPSC derived cardiomyocytes (CMs).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">genomic.annots</span><span class="op">$</span><span class="va">pcHiC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"pcHiC.CM.gr.rds"</span>, package <span class="op">=</span> <span class="st">"mapgen"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>You can use your own pcHiC data as follows, or skip this if you do not have relevant PCHi-C data.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">evapcHiC</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="va">pcHiC.file</span><span class="op">)</span>
<span class="va">pcHiC.gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_pcHiC.html">process_pcHiC</a></span><span class="op">(</span><span class="va">pcHiC</span><span class="op">)</span>
<span class="va">pcHiC.gr</span> <span class="op">&lt;-</span> <span class="va">pcHiC.gr</span><span class="op">[</span><span class="va">pcHiC.gr</span><span class="op">$</span><span class="va">gene_name</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">gene.annots</span><span class="op">$</span><span class="va">gene_name</span>,<span class="op">]</span> <span class="co"># restrict to protein coding genes</span>
<span class="va">genomic.annots</span><span class="op">$</span><span class="va">pcHiC</span> <span class="op">&lt;-</span> <span class="va">pcHiC.gr</span></code></pre></div>
<p>Add ABC scores from heart ventricle (from Nasser <em>et al.</em> <em>Nature</em> 2021). You may skip this if you do not have relevant ABC scores.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ABC</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"heart_ventricle-ENCODE_ABC.tsv.gz"</span>, package <span class="op">=</span> <span class="st">"mapgen"</span><span class="op">)</span><span class="op">)</span>
<span class="va">ABC.gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_ABC.html">process_ABC</a></span><span class="op">(</span><span class="va">ABC</span>, ABC.thresh <span class="op">=</span> <span class="fl">0.015</span>, full.element <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">ABC.gr</span> <span class="op">&lt;-</span> <span class="va">ABC.gr</span><span class="op">[</span><span class="va">ABC.gr</span><span class="op">$</span><span class="va">gene_name</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">gene.annots</span><span class="op">$</span><span class="va">gene_name</span>, <span class="op">]</span> <span class="co"># restrict to protein coding genes</span>
<span class="va">genomic.annots</span><span class="op">$</span><span class="va">ABC</span> <span class="op">&lt;-</span> <span class="va">ABC.gr</span></code></pre></div>
<p>Considering the fact that Hi-C and PC-HiC may miss contacts between close regions due to technical reasons, we also consider enhancer regions (the OCRs in this case) within 20 kb of active promoters as “enhancer loops”.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">genomic.annots</span><span class="op">$</span><span class="va">enhancer_regions</span> <span class="op">&lt;-</span> <span class="va">genomic.annots</span><span class="op">$</span><span class="va">OCRs_hg19</span><span class="op">[</span><span class="va">genomic.annots</span><span class="op">$</span><span class="va">OCRs_hg19</span><span class="op">$</span><span class="va">peakType</span><span class="op">!=</span><span class="st">"Promoter"</span>,<span class="op">]</span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">enhancer_nearby_promoter_20kb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/nearby_interactions.html">nearby_interactions</a></span><span class="op">(</span><span class="va">genomic.annots</span><span class="op">$</span><span class="va">enhancer_regions</span>, <span class="va">genomic.annots</span><span class="op">$</span><span class="va">active_promoters</span>, dist_limit <span class="op">=</span> <span class="fl">20000</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="run-gene-mapping">Run gene mapping<a class="anchor" aria-label="anchor" href="#run-gene-mapping"></a>
</h2>
<p>Run gene mapping using the following settings:</p>
<ul>
<li>enhancer.loop.method = “ABC.pcHiC.nearby20kb” (as explained above).</li>
<li>intron.mode = FALSE (do not assign intronic SNPs to the genes containing the introns).</li>
<li>c.dist = 50000 (distance weight parameter: 50kb).</li>
</ul>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gene.mapping.res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_gene_pip.html">compute_gene_pip</a></span><span class="op">(</span><span class="va">finemapstats.gr</span>, 
                                     <span class="va">genomic.annots</span>,
                                     enhancer.loop.method <span class="op">=</span> <span class="st">"ABC.pcHiC.nearby20kb"</span>,
                                     intron.mode <span class="op">=</span> <span class="cn">FALSE</span>,
                                     c.dist <span class="op">=</span> <span class="fl">50000</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Map SNPs to genes and assign weights ...</span>
<span class="co">## Include ABC scores in enhancer loops ... </span>
<span class="co">## Include pcHiC in enhancer loops ... </span>
<span class="co">## Include enhancers with nearby promoters (20kb) in enhancer loops ... </span>
<span class="co">## Assign SNPs in exons and active promoters ...</span>
<span class="co">## Assign SNPs in enhancers to linked genes through enhancer loops ...</span>
<span class="co">## Assign SNPs in enhancer regions to genes by distance weighting ...</span>
<span class="co">## Assign SNPs in UTRs (but not enhancer regions) to the UTR genes ... </span>
<span class="co">## Assign intergenic SNPs to genes by distance weighting ...</span>
<span class="co">## Compute gene PIP ...</span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gene.mapping.res</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##         snp chr       pos          pip     pval    zscore cs locus gene_name</span>
<span class="co">## 1 rs6658920   1  22257739 8.747860e-03  6.40472  5.057971  0    15     HSPG2</span>
<span class="co">## 2 rs6658920   1  22257739 8.747860e-03  6.40472  5.057971  0    15     USP48</span>
<span class="co">## 3  rs213734   1  41575699 5.171119e-05  3.56067 -3.611940  0    25     SCMH1</span>
<span class="co">## 4 rs4845618   1 154400015 1.149990e-05 10.16240 -6.537313  0    77      IL6R</span>
<span class="co">## 5 rs6687726   1 154400320 1.392576e-05 10.29390 -6.567164  0    77      IL6R</span>
<span class="co">## 6 rs2793374   1 205647508 1.950893e-04  4.70421  4.268657  0   106   SLC45A3</span>
<span class="co">##   category weight  frac_pip     gene_pip</span>
<span class="co">## 1      ABC      1 0.3333333 3.731341e-01</span>
<span class="co">## 2      ABC      1 0.3333333 2.591496e-02</span>
<span class="co">## 3      ABC      1 1.0000000 7.609949e-02</span>
<span class="co">## 4      ABC      1 1.0000000 3.297559e-05</span>
<span class="co">## 5      ABC      1 0.5000000 3.297559e-05</span>
<span class="co">## 6      ABC      1 1.0000000 4.721804e-02</span></code></pre>
<p>Extract gene-level result table</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gene.pip.res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extract_gene_level_result.html">extract_gene_level_result</a></span><span class="op">(</span><span class="va">gene.mapping.res</span>, <span class="va">gene.annots</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Extract gene level result ...</span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gene.pip.res</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##    chr       pos gene_name     gene_pip</span>
<span class="co">## 1 chr1  22263790     HSPG2 3.731341e-01</span>
<span class="co">## 2 chr1  22110099     USP48 2.591496e-02</span>
<span class="co">## 3 chr1  41707826     SCMH1 7.609949e-02</span>
<span class="co">## 4 chr1 154377669      IL6R 3.297559e-05</span>
<span class="co">## 5 chr1 205649587   SLC45A3 4.721804e-02</span>
<span class="co">## 6 chr1 205719404    NUCKS1 6.253132e-01</span></code></pre>
<p>Obtain 80% credible gene sets</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gene.cs.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gene_cs.html">gene_cs</a></span><span class="op">(</span><span class="va">gene.mapping.res</span>, by.locus <span class="op">=</span> <span class="cn">TRUE</span>, gene.cs.percent.thresh <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gene.cs.df</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 6 × 6</span></span>
<span class="co">##   locus gene_cs          gene_cs_locus_p… top_gene top_locus_gene_… top_gene_pip</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span>     7 CASZ1            CASZ1(0.886)     CASZ1               0.886        0.886</span>
<span class="co">## <span style="color: #BCBCBC;">2</span>    15 HSPG2,CELA3B,LD… HSPG2(0.373),CE… HSPG2               0.373        0.373</span>
<span class="co">## <span style="color: #BCBCBC;">3</span>    25 SLFNL1,CTPS1     SLFNL1(0.624),C… SLFNL1              0.624        0.624</span>
<span class="co">## <span style="color: #BCBCBC;">4</span>    31 BEND5            BEND5(0.997)     BEND5               0.997        0.997</span>
<span class="co">## <span style="color: #BCBCBC;">5</span>    32 FAF1,CDKN2C,C1o… FAF1(0.368),CDK… FAF1                0.368        0.390</span>
<span class="co">## <span style="color: #BCBCBC;">6</span>    33 TTC39A,EPS15     TTC39A(0.425),E… TTC39A              0.425        0.427</span></code></pre>
<p>Find the nearest gene for the top SNP in each locus</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">finemapstats.gr</span> <span class="op">&lt;-</span> <span class="va">finemapstats.gr</span><span class="op">[</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/setorder.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">finemapstats.gr</span><span class="op">$</span><span class="va">zscore</span><span class="op">)</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="op">]</span>
<span class="va">topsnps.gr</span> <span class="op">&lt;-</span> <span class="va">finemapstats.gr</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/duplicated.html" class="external-link">duplicated</a></span><span class="op">(</span><span class="va">finemapstats.gr</span><span class="op">$</span><span class="va">locus</span><span class="op">)</span>, <span class="op">]</span>
  
<span class="va">nearest_genes.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/find_nearest_genes.html">find_nearest_genes</a></span><span class="op">(</span><span class="va">topsnps.gr</span>, <span class="va">gene.annots</span>, dist.to <span class="op">=</span> <span class="st">"genebody"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="va">nearest_genes.df</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##          snp chr       pos nearest_gene</span>
<span class="co">## 1  rs2200732   4 111710042        PITX2</span>
<span class="co">## 2  rs2106261  16  73051620        ZFHX3</span>
<span class="co">## 3 rs11264280   1 154862952        KCNN3</span>
<span class="co">## 4 rs11598047  10 105342672        NEURL</span>
<span class="co">## 5 rs11773845   7 116191301         CAV1</span>
<span class="co">## 6   rs577676   1 170587340        PRRX1</span></code></pre>
<p>Make a Manhattan plot of the gene PIPs, and label genes with gene PIP &gt; 0.8.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/gene_manhattan_plot.html">gene_manhattan_plot</a></span><span class="op">(</span><span class="va">gene.pip.res</span>, sig.pip <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span></code></pre></div>
<p><img src="gene_mapping_tutorial_files/figure-html/gene-manhattan-plot-1.png" width="768"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kaixuan Luo, Alan Selewa, Sebastian Pott, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
