<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Tutorial for making track plots • mapgen</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Tutorial for making track plots"><meta property="og:description" content="mapgen"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mapgen</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.5.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/data_preparation_tutorial.html">Data preparation</a>
    </li>
    <li>
      <a href="../articles/enrichment_tutorial.html">Enrichment analysis</a>
    </li>
    <li>
      <a href="../articles/finemapping_tutorial.html">Functional fine-mapping</a>
    </li>
    <li>
      <a href="../articles/partition_pip_tutorial.html">Partition fine-mapping PIPs by annotation categories</a>
    </li>
    <li>
      <a href="../articles/gene_mapping_tutorial.html">Gene mapping</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/xinhe-lab/mapgen/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>





<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Tutorial for making track plots</h1>
                        <h4 data-toc-skip class="author">Kaixuan
Luo</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/mapgen/blob/HEAD/vignettes/make_trackplots.Rmd" class="external-link"><code>vignettes/make_trackplots.Rmd</code></a></small>
      <div class="hidden name"><code>make_trackplots.Rmd</code></div>

    </div>

    
    
<div id="input-data" class="section level2">
<h2>Input data</h2>
<ul>
<li>Fine-mapping summary statistics.</li>
<li>Gene annotations.</li>
<li>Functional annotation data, e.g.: chromatin loops, open chromatin
regions, histone modification data, etc.</li>
</ul>
<p>Please have the following Bioconductor packages installed:
<code>GenomicFeatures</code>, <code>rtracklayer</code>,
<code>Gviz</code>, <code>GenomicInteractions</code>,
<code>AnnotationDbi</code>, <code>org.Hs.eg.db</code>.</p>
<p>Load packages</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(GenomicFeatures) <span class="co">#  Making and manipulating annotations</span></span></code></pre></div>
<pre><code>## Warning: package &#39;GenomicFeatures&#39; was built under R version 4.2.2</code></pre>
<pre><code>## Warning: package &#39;GenomeInfoDb&#39; was built under R version 4.2.2</code></pre>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">library</span>(rtracklayer)</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="fu">library</span>(Gviz) <span class="co"># R package used to visualize track plots</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="fu">library</span>(GenomicInteractions) <span class="co"># visualize HiC loops</span></span></code></pre></div>
<pre><code>## Warning: package &#39;InteractionSet&#39; was built under R version 4.2.2</code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">library</span>(AnnotationDbi) <span class="co"># match gene ID to gene symbol</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="fu">library</span>(org.Hs.eg.db) <span class="co"># match gene ID to gene symbol</span></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="fu">library</span>(mapgen)</span></code></pre></div>
<p>Load fine-mapping summary statistics from the AFib study</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>finemapstats <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;AF.finemapping.sumstats.rds&quot;</span>, <span class="at">package =</span> <span class="st">&quot;mapgen&quot;</span>))</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>finemapstats <span class="ot">&lt;-</span> <span class="fu">process_finemapping_sumstats</span>(finemapstats, </span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>                                             <span class="at">snp =</span> <span class="st">&#39;snp&#39;</span>, </span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>                                             <span class="at">chr =</span> <span class="st">&#39;chr&#39;</span>, </span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>                                             <span class="at">pos =</span> <span class="st">&#39;pos&#39;</span>, </span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a>                                             <span class="at">pip =</span> <span class="st">&#39;susie_pip&#39;</span>, </span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a>                                             <span class="at">pval =</span> <span class="st">&#39;pval&#39;</span>, </span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a>                                             <span class="at">zscore =</span> <span class="st">&#39;zscore&#39;</span>, </span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a>                                             <span class="at">cs =</span> <span class="st">&#39;cs&#39;</span>, </span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a>                                             <span class="at">locus =</span> <span class="st">&#39;locus&#39;</span>,  </span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a>                                             <span class="at">pip.thresh =</span> <span class="dv">0</span>)</span></code></pre></div>
<pre><code>## Processing fine-mapping summary statistics ...</code></pre>
<p>Load genomic annotations (hg19) included in the package</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>genomic.annots <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;genomic.annots.hg19.rds&quot;</span>, <span class="at">package =</span> <span class="st">&quot;mapgen&quot;</span>))</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="fu">summary</span>(genomic.annots)</span></code></pre></div>
<pre><code>##           Length Class   Mode
## exons     195913 GRanges S4  
## introns   189762 GRanges S4  
## UTRs       50889 GRanges S4  
## promoters  19430 GRanges S4  
## genes      19430 GRanges S4</code></pre>
<p>gene annotations</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a>gene.annots <span class="ot">&lt;-</span> genomic.annots<span class="sc">$</span>genes</span></code></pre></div>
<p>Load promoter-capture HiC (PC-HiC) data from cardiomyocytes
(CMs).</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a>pcHiC <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;pcHiC.CM.gr.rds&quot;</span>, <span class="at">package =</span> <span class="st">&quot;mapgen&quot;</span>))</span></code></pre></div>
<p>Load ABC scores from heart ventricle (from Nasser <em>et al.</em>
<em>Nature</em> 2021).</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a>ABC <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;ABC.heart.ventricle.ENCODE.rds&quot;</span>, <span class="at">package =</span> <span class="st">&quot;mapgen&quot;</span>))</span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a><span class="co"># process_ABC() process a data frame of ABC scores and converts to a GRanges object.</span></span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>ABC <span class="ot">&lt;-</span> <span class="fu">process_ABC</span>(ABC, <span class="at">full.element =</span> <span class="cn">TRUE</span>)</span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a>ABC <span class="ot">&lt;-</span> ABC[ABC<span class="sc">$</span>gene_name <span class="sc">%in%</span> gene.annots<span class="sc">$</span>gene_name, ] <span class="co"># restrict to protein coding genes</span></span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a>ABC<span class="sc">$</span>promoter_chr <span class="ot">&lt;-</span> <span class="fu">seqnames</span>(ABC)</span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a>ABC<span class="sc">$</span>promoter_start <span class="ot">&lt;-</span> ABC<span class="sc">$</span>TargetGeneTSS</span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a>ABC<span class="sc">$</span>promoter_end <span class="ot">&lt;-</span> ABC<span class="sc">$</span>TargetGeneTSS</span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a>ABC<span class="sc">$</span>score <span class="ot">&lt;-</span> ABC<span class="sc">$</span>ABC.Score <span class="sc">*</span> <span class="dv">100</span></span></code></pre></div>
<p>Load H3K27ac and DHS <code>.bed</code> files.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a>trackdata.dir <span class="ot">&lt;-</span> <span class="st">&quot;../../mapgen_tutorial_data/trackplot/&quot;</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a>H3K27ac_peaks <span class="ot">&lt;-</span> rtracklayer<span class="sc">::</span><span class="fu">import</span>(<span class="fu">file.path</span>(trackdata.dir, <span class="st">&quot;H3K27ac.heart.concat.hg19.bed.gz&quot;</span>))</span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a>DHS_peaks <span class="ot">&lt;-</span> rtracklayer<span class="sc">::</span><span class="fu">import</span>(<span class="fu">file.path</span>(trackdata.dir, <span class="st">&quot;FetalHeart_E083.DNase.hg19.narrowPeak.bed.gz&quot;</span>))</span></code></pre></div>
<p>Load ATAC-seq counts data (the data should be in <code>wig</code>,
<code>bigWig/bw</code>, <code>bedGraph</code>, or <code>bam</code>
format)</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a>CM_counts <span class="ot">&lt;-</span> rtracklayer<span class="sc">::</span><span class="fu">import</span>(<span class="fu">file.path</span>(trackdata.dir, <span class="st">&quot;Cardiomyocyte.atac.hg19.bedGraph.gz&quot;</span>))</span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a>Endo_counts <span class="ot">&lt;-</span> rtracklayer<span class="sc">::</span><span class="fu">import</span>(<span class="fu">file.path</span>(trackdata.dir, <span class="st">&quot;Endothelial.atac.hg19.bedGraph.gz&quot;</span>)) </span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a>Fibro_counts <span class="ot">&lt;-</span> rtracklayer<span class="sc">::</span><span class="fu">import</span>(<span class="fu">file.path</span>(trackdata.dir, <span class="st">&quot;Fibroblast.atac.hg19.bedGraph.gz&quot;</span>))</span></code></pre></div>
<p>You can build a <code>TxDb</code> database (“.sqlite”) using gene
annotations (GTF format) from <a
href="https://www.gencodegenes.org">GENCODE</a>,</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a>txdb_gtf <span class="ot">&lt;-</span> <span class="fu">makeTxDbFromGFF</span>(<span class="fu">file.path</span>(trackdata.dir, <span class="st">&#39;gencode.v19.annotation.gtf.gz&#39;</span>), <span class="at">format =</span> <span class="st">&quot;gtf&quot;</span>)</span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a><span class="fu">saveDb</span>(txdb_gtf, <span class="fu">file.path</span>(trackdata.dir, <span class="st">&quot;gencode.v19.annotation.gtf.sqlite&quot;</span>))</span></code></pre></div>
<p>We can use to use the <code>TxDb</code> database.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a>txdb <span class="ot">&lt;-</span> <span class="fu">loadDb</span>(<span class="fu">file.path</span>(trackdata.dir, <span class="st">&quot;gencode.v19.annotation.gtf.sqlite&quot;</span>))</span></code></pre></div>
<p>If you are in Xin He lab at UChicago, you can access the gene
annotations and <code>TxDb</code> database in the directory:
<code>/project2/xinhe/shared_data/gencode/</code> from RCC.</p>
<p>You will need the <code>bigsnpr</code> package if you want to
visualize R^2 between SNPs using a reference panel in
<code>bigSNP</code> object.</p>
<p>We provided a <code>bigSNP</code> object of the reference genotype
panel from the 1000 Genomes (1KG) European population.</p>
<p>If you are in the He lab at UChicago, you can load the
<code>bigSNP</code> object from RCC as below.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="fu">library</span>(bigsnpr) <span class="co"># loading reference genotype for LD calculation</span></span></code></pre></div>
<pre><code>## Loading required package: bigstatsr</code></pre>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a>bigSNP <span class="ot">&lt;-</span> <span class="fu">snp_attach</span>(<span class="at">rdsfile =</span> <span class="st">&#39;/project2/xinhe/1kg/bigsnpr/EUR_variable_1kg.rds&#39;</span>)</span></code></pre></div>
</div>
<div id="make-track-plots" class="section level2">
<h2>Make track plots</h2>
<p>Plot CALU locus in the genomic region “chr7:128375000-128425000”
Highlight topSNP “rs55985730”</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="st">&quot;CM&quot;</span> <span class="ot">=</span> CM_counts, <span class="st">&quot;Endo&quot;</span> <span class="ot">=</span> Endo_counts, <span class="st">&quot;Fibro&quot;</span> <span class="ot">=</span> Fibro_counts)</span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a>peaks <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="st">&quot;H3K27ac&quot;</span> <span class="ot">=</span> H3K27ac_peaks, <span class="st">&quot;DHS&quot;</span> <span class="ot">=</span> DHS_peaks)</span>
<span id="cb22-3"><a href="#cb22-3" tabindex="-1"></a>loops <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="st">&quot;PC-HiC&quot;</span> <span class="ot">=</span> pcHiC)</span>
<span id="cb22-4"><a href="#cb22-4" tabindex="-1"></a></span>
<span id="cb22-5"><a href="#cb22-5" tabindex="-1"></a><span class="fu">track_plot</span>(finemapstats,</span>
<span id="cb22-6"><a href="#cb22-6" tabindex="-1"></a>           <span class="at">region =</span> <span class="st">&quot;chr7:128375000-128425000&quot;</span>,</span>
<span id="cb22-7"><a href="#cb22-7" tabindex="-1"></a>           gene.annots,</span>
<span id="cb22-8"><a href="#cb22-8" tabindex="-1"></a>           <span class="at">bigSNP =</span> bigSNP,</span>
<span id="cb22-9"><a href="#cb22-9" tabindex="-1"></a>           <span class="at">txdb =</span> txdb,</span>
<span id="cb22-10"><a href="#cb22-10" tabindex="-1"></a>           <span class="at">counts =</span> counts,</span>
<span id="cb22-11"><a href="#cb22-11" tabindex="-1"></a>           <span class="at">peaks =</span> peaks,</span>
<span id="cb22-12"><a href="#cb22-12" tabindex="-1"></a>           <span class="at">loops =</span> loops,</span>
<span id="cb22-13"><a href="#cb22-13" tabindex="-1"></a>           <span class="at">genome =</span> <span class="st">&quot;hg19&quot;</span>,</span>
<span id="cb22-14"><a href="#cb22-14" tabindex="-1"></a>           <span class="at">filter_loop_genes =</span> <span class="st">&quot;CALU&quot;</span>,</span>
<span id="cb22-15"><a href="#cb22-15" tabindex="-1"></a>           <span class="at">highlight_snps =</span> <span class="st">&quot;topSNP&quot;</span>, </span>
<span id="cb22-16"><a href="#cb22-16" tabindex="-1"></a>           <span class="at">counts.color =</span> <span class="fu">c</span>(<span class="st">&quot;red&quot;</span>, <span class="st">&quot;green&quot;</span>, <span class="st">&quot;purple&quot;</span>),</span>
<span id="cb22-17"><a href="#cb22-17" tabindex="-1"></a>           <span class="at">peaks.color =</span> <span class="fu">c</span>(<span class="st">&quot;navy&quot;</span>, <span class="st">&quot;blue&quot;</span>),</span>
<span id="cb22-18"><a href="#cb22-18" tabindex="-1"></a>           <span class="at">loops.color =</span> <span class="st">&quot;gray&quot;</span>, </span>
<span id="cb22-19"><a href="#cb22-19" tabindex="-1"></a>           <span class="at">rotation.title =</span> <span class="dv">0</span>,</span>
<span id="cb22-20"><a href="#cb22-20" tabindex="-1"></a>           <span class="at">genelabel.side =</span> <span class="st">&quot;left&quot;</span>,</span>
<span id="cb22-21"><a href="#cb22-21" tabindex="-1"></a>           <span class="at">verbose =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## 219 snps included.
## Color SNPs in PIP track by loci. 
## Adding CM track...
## Adding Endo track...
## Adding Fibro track...
## Adding H3K27ac track...
## Adding DHS track...
## Adding PC-HiC track...
## Only show PC-HiC loops linked to gene: CALU 
## Making gene track object using txdb database ...
## Highlight SNP: rs55985730 
## Making track plot ...</code></pre>
<p><img src="make_trackplots_files/figure-html/make-CALU-trackplot-1.png" width="1152" /></p>
<p>Plot HCN4 locus in the genomic region “chr15:73610000-73700000”
Highlight topSNP “rs7172038”</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="st">&quot;CM&quot;</span> <span class="ot">=</span> CM_counts, <span class="st">&quot;Endo&quot;</span> <span class="ot">=</span> Endo_counts, <span class="st">&quot;Fibro&quot;</span> <span class="ot">=</span> Fibro_counts)</span>
<span id="cb24-2"><a href="#cb24-2" tabindex="-1"></a>peaks <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="st">&quot;H3K27ac&quot;</span> <span class="ot">=</span> H3K27ac_peaks, <span class="st">&quot;DHS&quot;</span> <span class="ot">=</span> DHS_peaks)</span>
<span id="cb24-3"><a href="#cb24-3" tabindex="-1"></a>loops <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="st">&quot;ABC&quot;</span> <span class="ot">=</span> ABC)</span>
<span id="cb24-4"><a href="#cb24-4" tabindex="-1"></a></span>
<span id="cb24-5"><a href="#cb24-5" tabindex="-1"></a><span class="fu">track_plot</span>(finemapstats,</span>
<span id="cb24-6"><a href="#cb24-6" tabindex="-1"></a>           <span class="at">region =</span> <span class="st">&quot;chr15:73610000-73700000&quot;</span>,</span>
<span id="cb24-7"><a href="#cb24-7" tabindex="-1"></a>           gene.annots,</span>
<span id="cb24-8"><a href="#cb24-8" tabindex="-1"></a>           <span class="at">bigSNP =</span> bigSNP,</span>
<span id="cb24-9"><a href="#cb24-9" tabindex="-1"></a>           <span class="at">txdb =</span> txdb,</span>
<span id="cb24-10"><a href="#cb24-10" tabindex="-1"></a>           <span class="at">counts =</span> counts,</span>
<span id="cb24-11"><a href="#cb24-11" tabindex="-1"></a>           <span class="at">peaks =</span> peaks,</span>
<span id="cb24-12"><a href="#cb24-12" tabindex="-1"></a>           <span class="at">loops =</span> loops,</span>
<span id="cb24-13"><a href="#cb24-13" tabindex="-1"></a>           <span class="at">genome =</span> <span class="st">&quot;hg19&quot;</span>,</span>
<span id="cb24-14"><a href="#cb24-14" tabindex="-1"></a>           <span class="at">filter_loop_genes =</span> <span class="st">&quot;HCN4&quot;</span>,</span>
<span id="cb24-15"><a href="#cb24-15" tabindex="-1"></a>           <span class="at">highlight_snps =</span> <span class="st">&quot;topSNP&quot;</span>, </span>
<span id="cb24-16"><a href="#cb24-16" tabindex="-1"></a>           <span class="at">counts.color =</span> <span class="fu">c</span>(<span class="st">&quot;red&quot;</span>, <span class="st">&quot;green&quot;</span>, <span class="st">&quot;purple&quot;</span>),</span>
<span id="cb24-17"><a href="#cb24-17" tabindex="-1"></a>           <span class="at">peaks.color =</span> <span class="fu">c</span>(<span class="st">&quot;navy&quot;</span>, <span class="st">&quot;blue&quot;</span>),</span>
<span id="cb24-18"><a href="#cb24-18" tabindex="-1"></a>           <span class="at">loops.color =</span> <span class="st">&quot;gray&quot;</span>, </span>
<span id="cb24-19"><a href="#cb24-19" tabindex="-1"></a>           <span class="at">rotation.title =</span> <span class="dv">0</span>,</span>
<span id="cb24-20"><a href="#cb24-20" tabindex="-1"></a>           <span class="at">genelabel.side =</span> <span class="st">&quot;above&quot;</span>,</span>
<span id="cb24-21"><a href="#cb24-21" tabindex="-1"></a>           <span class="at">verbose =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## 463 snps included.
## Color SNPs in PIP track by loci. 
## Adding CM track...
## Adding Endo track...
## Adding Fibro track...
## Adding H3K27ac track...
## Adding DHS track...
## Adding ABC track...
## Only show ABC loops linked to gene: HCN4 
## Making gene track object using txdb database ...
## Highlight SNP: rs7172038 
## Making track plot ...</code></pre>
<p><img src="make_trackplots_files/figure-html/make-HCN4-trackplot-1.png" width="1152" /></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
      </div>

</div>



      <footer><div class="copyright">
  <p></p><p>Developed by Kaixuan Luo, Alan Selewa, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>
