<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Partition fine-mapping PIPs by annotation categories â€¢ mapgen</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Partition fine-mapping PIPs by annotation categories">
<meta property="og:description" content="mapgen">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mapgen</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.11</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/enrichment_finemapping_tutorial.html">Functional enrichment and fine-mapping</a>
    </li>
    <li>
      <a href="../articles/partition_pip_tutorial.html">Partition fine-mapping PIPs by annotation categories</a>
    </li>
    <li>
      <a href="../articles/gene_mapping_tutorial.html">Gene mapping</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Partition fine-mapping PIPs by annotation
categories</h1>
                        <h4 data-toc-skip class="author">Kaixuan
Luo</h4>
            
      
      
      <div class="hidden name"><code>partition_pip_tutorial.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="partitioning-pips-by-cell-types-and-annotation-categories">Partitioning PIPs by cell types and annotation categories<a class="anchor" aria-label="anchor" href="#partitioning-pips-by-cell-types-and-annotation-categories"></a>
</h2>
<p>Here, we show an example using data from our heart single-cell study.
We assigned the likely cell type(s) through which the causal variants
act in each locus using fine-mapped SNPs, and cell-type specific open
chromatin regions (OCRs).</p>
<p><strong>Required input data</strong>:</p>
<ul>
<li>Fine-mapping summary statistics.</li>
<li>A list of GRanges objects of annotation regions.</li>
</ul>
<p>Load R packages</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">mapgen</span><span class="op">)</span></span></code></pre></div>
<p>Load fine-mapping summary statistics.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">finemapstats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"aFib_Finemapped.tble.rds"</span>, package <span class="op">=</span> <span class="st">"mapgen"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">finemapstats.gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_finemapping_sumstats.html">process_finemapping_sumstats</a></span><span class="op">(</span><span class="va">finemapstats</span>, </span>
<span>                                                snp <span class="op">=</span> <span class="st">'snp'</span>, chr <span class="op">=</span> <span class="st">'chr'</span>, </span>
<span>                                                pos <span class="op">=</span> <span class="st">'pos'</span>, pip <span class="op">=</span> <span class="st">'susie_pip'</span>, </span>
<span>                                                pval <span class="op">=</span> <span class="st">'pval'</span>, zscore <span class="op">=</span> <span class="st">'zscore'</span>, </span>
<span>                                                cs <span class="op">=</span> <span class="st">'CS'</span>, locus <span class="op">=</span> <span class="st">'locus'</span>,  </span>
<span>                                                pip.thresh <span class="op">=</span> <span class="fl">1e-5</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co"># Processing fine-mapping summary statistics ...</span></span>
<span><span class="co"># Filter SNPs with PIP threshold of 1e-05</span></span></code></pre>
<p>We can partition PIPs into different functional annotation
categories.</p>
<p>Load genomic annotations (hg19).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genomic.annots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"genomic.annots.hg19.gr.rds"</span>, package <span class="op">=</span> <span class="st">"mapgen"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Load all OCRs (hg19).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genomic.annots</span><span class="op">$</span><span class="va">OCRs_hg19</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"OCRs.hg19.gr.rds"</span>, package <span class="op">=</span> <span class="st">"mapgen"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Create a list of the annotations, with priority in the order of OCRs,
UTRS, Exons, and Introns.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">annots.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>OCRs <span class="op">=</span> <span class="va">genomic.annots</span><span class="op">$</span><span class="va">OCRs_hg19</span>,</span>
<span>                    UTRs <span class="op">=</span> <span class="va">genomic.annots</span><span class="op">$</span><span class="va">UTRs</span>,</span>
<span>                    Exons <span class="op">=</span> <span class="va">genomic.annots</span><span class="op">$</span><span class="va">exons</span>,</span>
<span>                    Introns <span class="op">=</span> <span class="va">genomic.annots</span><span class="op">$</span><span class="va">introns</span><span class="op">)</span></span></code></pre></div>
<p>Sum PIPs within annotation categories. Unlike
partition_pip_regions(), it is OK to have overlapping annotations here.
If a SNP is in multiple annotation categories, it will be assigned to
the first ordered category.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sum_pip_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/partition_pip_annots.html">partition_pip_annots</a></span><span class="op">(</span><span class="va">finemapstats.gr</span>, <span class="va">annots.list</span><span class="op">)</span></span></code></pre></div>
<p>Sum of PIPs in each annotation category:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sum.pips</span> <span class="op">&lt;-</span> <span class="va">sum_pip_res</span><span class="op">$</span><span class="va">sum.pips</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sum.pips</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#            OCRs        UTRs        Exons   Introns       others</span></span>
<span><span class="co"># 7  7.680600e-01 0.000000000 0.000000e+00 0.2318394 0.0000000000</span></span>
<span><span class="co"># 15 4.596813e-01 0.000000000 0.000000e+00 0.4912706 0.0489539561</span></span>
<span><span class="co"># 25 4.968126e-03 0.001024686 1.507626e-05 0.9913449 0.0008067977</span></span>
<span><span class="co"># 31 1.404386e-05 0.000000000 0.000000e+00 0.9996138 0.0000000000</span></span>
<span><span class="co"># 32 5.370896e-01 0.000000000 0.000000e+00 0.2069662 0.2559185487</span></span>
<span><span class="co"># 33 3.466723e-01 0.037586895 0.000000e+00 0.4991703 0.1165574597</span></span></code></pre>
<p>Number of SNPs included in each annotation category:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n.snps</span> <span class="op">&lt;-</span> <span class="va">sum_pip_res</span><span class="op">$</span><span class="va">n.snps</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">n.snps</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#    OCRs UTRs Exons Introns others</span></span>
<span><span class="co"># 7     8    0     0      14      0</span></span>
<span><span class="co"># 15   23    0     0      15     31</span></span>
<span><span class="co"># 25   20    2     1      93     17</span></span>
<span><span class="co"># 31    1    0     0      11      0</span></span>
<span><span class="co"># 32    2    0     0      39     36</span></span>
<span><span class="co"># 33   18    1     0     144     30</span></span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sum.pips</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sum.pips</span><span class="op">)</span> <span class="op">==</span> <span class="st">"others"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Intergenic"</span></span>
<span><span class="va">sum.pips</span> <span class="op">&lt;-</span> <span class="va">sum.pips</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">sum.pips</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">OCRs</span>, <span class="va">UTRs</span>, <span class="va">Exons</span>, <span class="va">Introns</span>, <span class="va">Intergenic</span><span class="op">)</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">prop.pip.mat</span> <span class="op">&lt;-</span> <span class="va">sum.pips</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">sum.pips</span><span class="op">)</span></span></code></pre></div>
<p>We can make a structure plot to show the proportion of PIPs in each
annotation category.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">categories</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"OCRs"</span>, <span class="st">"UTRs"</span>, <span class="st">"Exons"</span>, <span class="st">"Introns"</span>, <span class="st">"Intergenic"</span><span class="op">)</span></span>
<span><span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>OCRs <span class="op">=</span> <span class="st">"#E18727FF"</span>, UTRs <span class="op">=</span> <span class="st">"#238b45"</span>, Exons <span class="op">=</span>  <span class="st">"#bee6af"</span>, Introns <span class="op">=</span> <span class="st">"#B09C85FF"</span>, Intergenic <span class="op">=</span> <span class="st">"#aaaaaa"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compile_structure_plot_data.html">compile_structure_plot_data</a></span><span class="op">(</span><span class="va">prop.pip.mat</span>, categories <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">rev</a></span><span class="op">(</span><span class="va">categories</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/structure_plot.html">structure_plot</a></span><span class="op">(</span><span class="va">dat</span>, colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">rev</a></span><span class="op">(</span><span class="va">colors</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="partition_pip_tutorial_files/figure-html/structure-plot-annots-1.png" width="960">
We can further partition PIPs into disjoint OCRs for different cell
types.</p>
<p>Load cell type OCR peaks. This is a list of GRanges objects
containing disjoint OCRs for different cell types.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">disjoint_OCRs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"disjoint_OCRs_hg19.grlist.rds"</span>, package <span class="op">=</span> <span class="st">"mapgen"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Sum PIPs within cell-type specific OCRs.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sum_pip_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/partition_pip_regions.html">partition_pip_regions</a></span><span class="op">(</span><span class="va">finemapstats.gr</span>, <span class="va">disjoint_OCRs</span><span class="op">)</span></span></code></pre></div>
<p>Sum of PIPs in each cell type OCR category:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sum.pips</span> <span class="op">&lt;-</span> <span class="va">sum_pip_res</span><span class="op">$</span><span class="va">sum.pips</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sum.pips</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#    Cardiomyocyte Endothelial Fibroblast Lymphoid Myeloid Pericyte   Shared 2-3</span></span>
<span><span class="co"># 7   7.663085e-01 0.000000000 0.00000000        0       0        0 0.0000000000</span></span>
<span><span class="co"># 15  2.173871e-01 0.004280965 0.08199708        0       0        0 0.0002198158</span></span>
<span><span class="co"># 25  1.417186e-04 0.000000000 0.00000000        0       0        0 0.0037536142</span></span>
<span><span class="co"># 31  1.404386e-05 0.000000000 0.00000000        0       0        0 0.0000000000</span></span>
<span><span class="co"># 32  5.367133e-01 0.000000000 0.00000000        0       0        0 0.0000000000</span></span>
<span><span class="co"># 33  7.847140e-02 0.000000000 0.00000000        0       0        0 0.0015118886</span></span>
<span><span class="co">#       Shared 4+</span></span>
<span><span class="co"># 7  0.000000e+00</span></span>
<span><span class="co"># 15 2.485050e-02</span></span>
<span><span class="co"># 25 3.452405e-04</span></span>
<span><span class="co"># 31 0.000000e+00</span></span>
<span><span class="co"># 32 3.762834e-04</span></span>
<span><span class="co"># 33 8.633893e-05</span></span></code></pre>
<p>Filter loci with a cutoff of total PIPs in OCR &gt; 0.25, compute the
proportion of PIPs partitioned in each cell type category.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sum.pips.filtered</span> <span class="op">&lt;-</span> <span class="va">sum.pips</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">sum.pips</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0.25</span>,<span class="op">]</span></span>
<span><span class="va">prop.pip.mat</span> <span class="op">&lt;-</span> <span class="va">sum.pips.filtered</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">sum.pips.filtered</span><span class="op">)</span></span>
<span><span class="va">prop.pip.mat</span> <span class="op">&lt;-</span> <span class="va">prop.pip.mat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">prop.pip.mat</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">Cardiomyocyte</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span></span></code></pre></div>
<p>We can make a structure plot to show the proportion of PIPs in each
cell type category.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">categories</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Cardiomyocyte"</span>, <span class="st">"Endothelial"</span>, <span class="st">"Fibroblast"</span>, <span class="st">"Lymphoid"</span>, </span>
<span>                <span class="st">"Myeloid"</span>, <span class="st">"Pericyte"</span>, <span class="st">"Shared 2-3"</span>, <span class="st">"Shared 4+"</span><span class="op">)</span></span>
<span><span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#b22222"</span>, <span class="st">"#8DD3C7"</span>, <span class="st">"#BEBADA"</span>, <span class="st">"#FB8072"</span>, </span>
<span>            <span class="st">"#80B1D3"</span>, <span class="st">"#B3DE69"</span>, <span class="st">"royalblue"</span>, <span class="st">"#003C86"</span><span class="op">)</span></span>
<span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compile_structure_plot_data.html">compile_structure_plot_data</a></span><span class="op">(</span><span class="va">prop.pip.mat</span>, categories <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">rev</a></span><span class="op">(</span><span class="va">categories</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/structure_plot.html">structure_plot</a></span><span class="op">(</span><span class="va">dat</span>, colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">rev</a></span><span class="op">(</span><span class="va">colors</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="partition_pip_tutorial_files/figure-html/structure-plot-OCRs-1.png" width="960"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kaixuan Luo, Alan Selewa, Sebastian Pott, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
