<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ Mapgen</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Mapgen</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/enrichment_finemapping_tutorial.html">Functional enrichment and fine mapping tutorial</a>
    </li>
    <li>
      <a href="../articles/gene_mapping_tutorial.html">Gene mapping tutorial</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="LDblock_view_summary.html">LDblock_view_summary()</a></code> </p>
        </td>
        <td><p>LD block view summary table</p></td>
      </tr><tr><td>
          <p><code><a href="PrepareSusieData.html">PrepareSusieData()</a></code> </p>
        </td>
        <td><p>PrepareSusieData</p></td>
      </tr><tr><td>
          <p><code><a href="PrepareTorusFiles.html">PrepareTorusFiles()</a></code> </p>
        </td>
        <td><p>PrepareTorusFiles</p></td>
      </tr><tr><td>
          <p><code><a href="RunCleaner.html">RunCleaner()</a></code> </p>
        </td>
        <td><p>RunCleaner</p></td>
      </tr><tr><td>
          <p><code><a href="RunFinemapping.html">RunFinemapping()</a></code> </p>
        </td>
        <td><p>RunFinemapping</p></td>
      </tr><tr><td>
          <p><code><a href="RunTorus.html">RunTorus()</a></code> </p>
        </td>
        <td><p>RunTorus</p></td>
      </tr><tr><td>
          <p><code><a href="RunTorusFDR.html">RunTorusFDR()</a></code> </p>
        </td>
        <td><p>RunTorusFDR</p></td>
      </tr><tr><td>
          <p><code><a href="annotator.html">annotator()</a></code> </p>
        </td>
        <td><p>Each annotation gets assigned SNPs based on overlap</p></td>
      </tr><tr><td>
          <p><code><a href="annotator_merged.html">annotator_merged()</a></code> </p>
        </td>
        <td><p>Annotations for causal SNPs (apply these after fine-mapping!)</p></td>
      </tr><tr><td>
          <p><code><a href="assign.locus.snp.html">assign.locus.snp()</a></code> </p>
        </td>
        <td><p>Assigns each SNP to one ld-block</p></td>
      </tr><tr><td>
          <p><code><a href="clean_sumstats.html">clean_sumstats()</a></code> </p>
        </td>
        <td><p>Clean summary statistics</p></td>
      </tr><tr><td>
          <p><code><a href="compute_gene_pip.html">compute_gene_pip()</a></code> </p>
        </td>
        <td><p>Map SNPs to genes and assign weights</p></td>
      </tr><tr><td>
          <p><code><a href="extract_gene_level_result.html">extract_gene_level_result()</a></code> </p>
        </td>
        <td><p>Extract gene-level result from SNP-level gene mapping result</p></td>
      </tr><tr><td>
          <p><code><a href="filter_coaccess.html">filter_coaccess()</a></code> </p>
        </td>
        <td><p>Filtering coaccessibility data</p></td>
      </tr><tr><td>
          <p><code><a href="find_nearest_genes.html">find_nearest_genes()</a></code> </p>
        </td>
        <td><p>Find nearest genes (by distance to TSS or gene body)</p></td>
      </tr><tr><td>
          <p><code><a href="gene_by_distance.html">gene_by_distance()</a></code> </p>
        </td>
        <td><p>Find all genes (promoters) within 1MB genes and assign weights based on distance</p></td>
      </tr><tr><td>
          <p><code><a href="gene_cs.html">gene_cs()</a></code> </p>
        </td>
        <td><p>Get gene credible sets from SNP level gene mapping table</p></td>
      </tr><tr><td>
          <p><code><a href="gene_manhattan_plot.html">gene_manhattan_plot()</a></code> </p>
        </td>
        <td><p>Make gene manhattan plot</p></td>
      </tr><tr><td>
          <p><code><a href="gene_view_summary.html">gene_view_summary()</a></code> </p>
        </td>
        <td><p>Gene view summary table</p></td>
      </tr><tr><td>
          <p><code><a href="load_narrowpeaks.html">load_narrowpeaks()</a></code> </p>
        </td>
        <td><p>Load ENCODE narrow peak data and convert to a GRanges object</p></td>
      </tr><tr><td>
          <p><code><a href="make_genomic_annots.html">make_genomic_annots()</a></code> </p>
        </td>
        <td><p>Make genomic annotations from GTF file</p></td>
      </tr><tr><td>
          <p><code><a href="merge_susie_sumstats.html">merge_susie_sumstats()</a></code> </p>
        </td>
        <td><p>merges susie results with original summary statistics data frame</p></td>
      </tr><tr><td>
          <p><code><a href="partition_pip_annots.html">partition_pip_annots()</a></code> </p>
        </td>
        <td><p>Partition PIPs into functional categories</p></td>
      </tr><tr><td>
          <p><code><a href="partition_pip_regions.html">partition_pip_regions()</a></code> </p>
        </td>
        <td><p>Partition PIPs into functional annotation regions</p></td>
      </tr><tr><td>
          <p><code><a href="process_ABC.html">process_ABC()</a></code> </p>
        </td>
        <td><p>Process ABC scores and save as a GRanges object</p></td>
      </tr><tr><td>
          <p><code><a href="process_finemapping_sumstat.html">process_finemapping_sumstat()</a></code> </p>
        </td>
        <td><p>Process fine mapping summary statistics data</p></td>
      </tr><tr><td>
          <p><code><a href="process_pcHiC.html">process_pcHiC()</a></code> </p>
        </td>
        <td><p>Process pcHiC data and save as a GRanges object</p></td>
      </tr><tr><td>
          <p><code><a href="run.susie.html">run.susie()</a></code> </p>
        </td>
        <td><p>run SUSIE</p></td>
      </tr><tr><td>
          <p><code><a href="snp_view_summary.html">snp_view_summary()</a></code> </p>
        </td>
        <td><p>SNP view summary table</p></td>
      </tr><tr><td>
          <p><code><a href="structure_plot.html">structure_plot()</a></code> </p>
        </td>
        <td><p>structure plot</p></td>
      </tr><tr><td>
          <p><code><a href="substract_exons.html">substract_exons()</a></code> </p>
        </td>
        <td><p>Substract all exon coordinates from all gene coordinates to get all intron coordinates</p></td>
      </tr><tr><td>
          <p><code><a href="sum_pip_annots.html">sum_pip_annots()</a></code> </p>
        </td>
        <td><p>Sum PIPs for each annotation category</p></td>
      </tr><tr><td>
          <p><code><a href="sum_pip_regions.html">sum_pip_regions()</a></code> </p>
        </td>
        <td><p>Sum PIPs for genomic regions</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Kaixuan Luo, Alan Selewa, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

      </footer></div>

  


  

  </body></html>

